<p><img src="logo.svg" width="100"></p>

# Tumor Mutational Burden (TMB) quantification

[![pipeline status](https://gitlab.com/acc-bioinfo/Clinical_utility_toolkit/tmbtool/badges/master/pipeline.svg)](https://gitlab.com/acc-bioinfo/Clinical_utility_toolkit/tmbtool/-/commits/master)




## Installation

External dependencies

* git lfs - [Install](https://git-lfs.github.com/)
* ImageMagick [Install](https://imagemagick.org/script/download.php)

Download the package with the git clone command:

```bash
git clone git@gitlab.com:acc-bioinfo/Clinical_utility_toolkit/tmbtool.git
# full all the large files
git lfs pull
```

Initiate R and install the R package using devtools (devtools needs to be installed as well)

```{r}
library("devtools")
devtools::install("TMBleR")
```

## Using TMBleR without installing it

TMBleR is also available as a docker container

```
# Auth into the gitlab register
docker login registry.gitlab.com
# Pull the latest docker image
docker pull registry.gitlab.com/acc-bioinfo/clinical_utility_toolkit/tmbtool:latest
# Run the container on port 8787
docker run \
    --name tmbler \
    --rm \
    -e USER=$(whoami) \
    -e PASSWORD=helloworld \
    -e USERID=$UID \
    -p 8787:8787 \
    -v $(pwd)/workdir \
    registry.gitlab.com/acc-bioinfo/clinical_utility_toolkit/tmbtool:latest
```

This will run a docker instance of TMBleR mounting your current directory into ```/workdir``` within the container, and 
opening port 8787 on your host machine, so that you can access RStudio interface through the browser by visiting 
http://localhost:8787 (username: your user, password: "helloworld")


## Repository structure
* ```README.md``` this file
* ```R``` - folder with code for functions
* ```data``` - folder with .rda objects containing COSMIC dataset or list of 
 human tumor suppressor genes loaded by functions or containing datasets which
 can be easily be loaded in the examples of the vignette by the function "load"
* ```data-raw```- folder with source files for .rda objects in data and with 
scripts used to generate them 
* ```Ã¬nst/extdata``` - folder containing VCF datasets and BED or TXT files for
sequencing design which can be read in input in the examples of the vignette as
raw data
* ```Output```- folder containing output generated by the examples from the 
vignette
* ```Vignettes```- folder containing the vignette
* ```man```- folder containing .Rd files generated by roxygen2 with help for each 
function

## Input
For TMB quantification
* a VCF file containing somatic SNV calls (mandatory)
* a file describing sequencing design: either a BED file with genomic coordinates 
of the targeted regions or a TXT file with official symbols of targeted genes (one 
per line) (mandatory)

For ROC curves of TMB performance in immunotherapy responders classification
* a dataframe containing TMB values (numeric vector, any name) and clinical response 
(factor with only the following two levels: 'responder', 'nonresponder'; mandatory 
name: "ClinicalResponse" )


