<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Tumor Mutational Burden Quantification From Gene Panels And WES • TMBleR</title>
<!-- favicons --><link rel="icon" type="image/png" sizes="16x16" href="favicon-16x16.png">
<link rel="icon" type="image/png" sizes="32x32" href="favicon-32x32.png">
<link rel="apple-touch-icon" type="image/png" sizes="180x180" href="apple-touch-icon.png">
<link rel="apple-touch-icon" type="image/png" sizes="120x120" href="apple-touch-icon-120x120.png">
<link rel="apple-touch-icon" type="image/png" sizes="76x76" href="apple-touch-icon-76x76.png">
<link rel="apple-touch-icon" type="image/png" sizes="60x60" href="apple-touch-icon-60x60.png">
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="bootstrap-toc.css">
<script src="bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="pkgdown.css" rel="stylesheet">
<script src="pkgdown.js"></script><meta property="og:title" content="Tumor Mutational Burden Quantification From Gene Panels And WES">
<meta property="og:description" content="The TMBleR R package allows to calculate TMB from gene panels with
    a custom combination of the most common parameters for variant filtering 
    adopted in literature (removal of known cancer variants, filtering by VAF, etc).
    It also allows to plot TMB distributions across several samples or based on
    different set of filtering parameters, to calculate and plot the correlation 
    between panel-based and WES-based TMB quantification, to simulate a gene panel 
    performance by downsampling a WES dataset, to build ROC curves to evaluate TMB 
    performance in predicting immunotherapy responders and to determine the optimal 
    TMB cutoff for immunotherapy responders identification.">
<meta property="og:image" content="/logo.svg">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    <div class="container template-home">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="index.html">TMBleR</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">0.1.1</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="index.html">
    <span class="fas fa-home fa-lg"></span>
     
  </a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Get Started
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="articles/Introduction_to_TMBleR.html">An introduction to TMBleR</a>
    </li>
    <li>
      <a href="articles/HOWTO_Import_external_data.html">HOWTO_Import_external_data</a>
    </li>
  </ul>
</li>
<li>
  <a href="reference/index.html">Reference</a>
</li>
<li>
  <a href="news/index.html">Changelog</a>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://hub.docker.com/repository/docker/aguida/tmbler">
    <span class="fab fa-docker fa-lg"></span>
     
    docker hub
  </a>
</li>
<li>
  <a href="https://github.com/acc-bioinfo/TMBleR">
    <span class="fa fa fa fa-github fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="https://bioserver.ieo.it/shiny/app/tmbler">
    <span class="fas fa-external-link-alt"></span>
     
    shiny app
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><div class="row">
  <div class="contents col-md-9">

<p style="text-align: center;">
<img src="logo.svg" width="150"></p>
<div id="tumor-mutational-burden-tmb-quantification" class="section level1">
<div class="page-header"><h1 class="hasAnchor">
<a href="#tumor-mutational-burden-tmb-quantification" class="anchor"></a>Tumor Mutational Burden (TMB) quantification</h1></div>
<p><strong>Estimate the Tumor Mutational Burden in genomics sequencing efforts like WES and genomic panels.</strong></p>
<p style="text-align: center;">
[ <a href="https://github.com/acc-bioinfo/TMBleR">Github</a> ] - [ <a href="https://acc-bioinfo.github.io/TMBleR/">Website</a> ] - [ <a href="https://hub.docker.com/repository/docker/aguida/tmbler">Docker Image</a> ] - [ <a href="https://bioserver.ieo.it/shiny/app/tmbler">Demo webapp</a> ]
</p>
<hr>
<p>TMB has been proposed as a predictive biomarker for immunotherapy response in cancer patients, as it is thought to enrich for tumors with high neoantigen load. TMB assessed by Whole Exome Sequencing (WES) is considered the gold standard but remains confined to research settings. Targeted enrichment panels of various genomic sizes are emerging as a more sustainable methodology for assessing TMB in the clinical setting. However, panel-based sequencing inevitably leads to a nonrandom sampling of the total population of potential neoantigens. Thus, the choice of variant classes to be included in TMB calculation (e.g. synonymous mutations, mutations in cancer driver genes, low-allelic frequency mutations) may have a major impact on TMB quantification, its predictive power, and therefore its clinical utility. Significant efforts are being conducted to provide a standardized framework for TMB calculation. TMBleR allows to streamline the TMB estimation process, to precisely measure the impact of different variant filtering options on TMB quantification and, more importantly, on its predictive power, by using published datasets of immunotherapy-treated patients with available outcome data.</p>
<p>We provide the community with a tool for maximizing the clinical impact of panel-based TMB measurement.</p>
<hr>
<div id="getting-started" class="section level2">
<h2 class="hasAnchor">
<a href="#getting-started" class="anchor"></a>Getting started</h2>
<p>TMBleR is available in different formats.</p>
<ul>
<li>
<strong>demo</strong>: If you only want to get a quick estimate of the TMB without having to install the software, have a look at the shiny web app available at the link <a href="https://bioserver.ieo.it/shiny/app/tmbler">https://bioserver.ieo.it/shiny/app/tmbler</a>. (<strong>Note</strong>, The app is not designed to support large computations with many samples).</li>
<li>As a <strong>Docker image</strong> ( See “Using TMBleR with no installation”)</li>
<li>As an <strong>R package</strong> (See “Install TMBleR as an R package”)</li>
</ul>
</div>
<div id="using-tmbler-with-no-installation" class="section level2">
<h2 class="hasAnchor">
<a href="#using-tmbler-with-no-installation" class="anchor"></a>Using TMBleR with no installation</h2>
<p>TMBleR is available as a docker container. This is the simplest way to get TMBleR to run</p>
<pre><code># Auth into the gitlab register
docker login docker.io
# Pull the latest docker image
docker pull aguida/tmbler:latest
# Run the container on port 8787
docker run \
    --name tmbler \
    --rm \
    -e USER=$(whoami) \
    -e PASSWORD=helloworld \
    -e USERID=$UID \
    -p 8787:8787 \
    -v $(pwd)/workdir \
    aguida/tmbler:latest</code></pre>
<p>This will run a docker instance of TMBleR mounting your current directory into <code>/workdir</code> within the container, and opening port 8787 on your host machine, so that you can access RStudio interface through the browser by visiting <a href="http://localhost:8787" class="uri">http://localhost:8787</a> (username: your user, password: “helloworld”)</p>
<p><strong>Note:</strong> By default the docker container will be initialized in the folder <code>/home/&lt;yourusername</code>. Use the Rstudio file manager to open <code>/workdir</code> or any other mounted folder of your choice.</p>
</div>
<div id="install-tmbler-as-an-r-package" class="section level2">
<h2 class="hasAnchor">
<a href="#install-tmbler-as-an-r-package" class="anchor"></a>Install TMBleR as an R package</h2>
<p>Prerequisites. TMBleR requires the following dependencies to be installed on your system:</p>
<ul>
<li>git lfs - <a href="https://git-lfs.github.com/">Install</a>
</li>
<li>ImageMagick <a href="https://imagemagick.org/script/download.php">Install</a>
</li>
</ul>
<p>Download the package with the git clone command:</p>
<div class="sourceCode" id="cb2"><pre class="sourceCode bash"><code class="sourceCode bash"><span id="cb2-1"><a href="#cb2-1" aria-hidden="true" tabindex="-1"></a><span class="fu">git</span> clone https://github.com/acc-bioinfo/TMBleR.git</span>
<span id="cb2-2"><a href="#cb2-2" aria-hidden="true" tabindex="-1"></a><span class="co"># full all the large files</span></span>
<span id="cb2-3"><a href="#cb2-3" aria-hidden="true" tabindex="-1"></a><span class="fu">git</span> -C TMBleR lfs pull</span></code></pre></div>
<p>Initiate R and install the R package using devtools (devtools needs to be installed as well)</p>
<pre class="{r}"><code><a href="https://devtools.r-lib.org/">library("devtools")
devtools::install("TMBleR")</a></code></pre>
</div>
<div id="usage" class="section level2">
<h2 class="hasAnchor">
<a href="#usage" class="anchor"></a>Usage</h2>
<p>To learn how to use TMBleR, please refer to the introductory vignette posted at this link:</p>
<ul>
<li><a href="https://acc-bioinfo.github.io/TMBleR/articles/Introduction_to_TMBleR.html">https://acc-bioinfo.github.io/TMBleR/articles/Introduction_to_TMBleR.html</a></li>
</ul>
</div>
</div>

  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <div class="license">
<h2>License</h2>
<ul class="list-unstyled">
<li><a href="https://www.r-project.org/Licenses/GPL-3">GPL-3</a></li>
</ul>
</div>
<div class="developers">
<h2>Developers</h2>
<ul class="list-unstyled">
<li>Alessandro Guida <br><small class="roles"> Author, maintainer </small>  </li>
<li>Laura Fancello <br><small class="roles"> Author </small>  </li>
<li>Luca Mazzarella <br><small class="roles"> Author </small>  </li>
<li><a href="authors.html">All authors...</a></li>
</ul>
</div>

  </div>
</div>


      <footer><div class="copyright">
  <p>Developed by Alessandro Guida, Laura Fancello, Luca Mazzarella.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.6.1.</p>
</div>

      </footer>
</div>

  


  </body>
</html>
