<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>An introduction to TMBleR • TMBleR</title>
<!-- favicons --><link rel="icon" type="image/png" sizes="16x16" href="../favicon-16x16.png">
<link rel="icon" type="image/png" sizes="32x32" href="../favicon-32x32.png">
<link rel="apple-touch-icon" type="image/png" sizes="180x180" href="../apple-touch-icon.png">
<link rel="apple-touch-icon" type="image/png" sizes="120x120" href="../apple-touch-icon-120x120.png">
<link rel="apple-touch-icon" type="image/png" sizes="76x76" href="../apple-touch-icon-76x76.png">
<link rel="apple-touch-icon" type="image/png" sizes="60x60" href="../apple-touch-icon-60x60.png">
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="An introduction to TMBleR">
<meta property="og:description" content="TMBleR">
<meta property="og:image" content="/logo.svg">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">TMBleR</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">0.1.0</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../index.html">
    <span class="fas fa-home fa-lg"></span>
     
  </a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Get Started
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/Introduction_to_TMBleR.html">An introduction to TMBleR</a>
    </li>
    <li>
      <a href="../articles/HOWTO_Import_external_data.html">HOWTO_Import_external_data</a>
    </li>
  </ul>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li>
  <a href="../news/index.html">Changelog</a>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://404.index">
    <span class="fab fa-docker fa-lg"></span>
     
    docker hub
  </a>
</li>
<li>
  <a href="https://gitlab.com/acc-bioinfo/Clinical_utility_toolkit/tmbtool">
    <span class="fa fa fa fa-gitlab fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="https://bioserver.ieo.it/shiny/app/tmbler">
    <span class="fas fa-external-link-alt"></span>
     
    shiny app
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><script src="Introduction_to_TMBleR_files/header-attrs-2.6/header-attrs.js"></script><script src="Introduction_to_TMBleR_files/htmlwidgets-1.5.3/htmlwidgets.js"></script><script src="Introduction_to_TMBleR_files/jquery-1.12.4/jquery.min.js"></script><link href="Introduction_to_TMBleR_files/datatables-css-0.0.0/datatables-crosstalk.css" rel="stylesheet">
<script src="Introduction_to_TMBleR_files/datatables-binding-0.13/datatables.js"></script><link href="Introduction_to_TMBleR_files/dt-core-1.10.20/css/jquery.dataTables.min.css" rel="stylesheet">
<link href="Introduction_to_TMBleR_files/dt-core-1.10.20/css/jquery.dataTables.extra.css" rel="stylesheet">
<script src="Introduction_to_TMBleR_files/dt-core-1.10.20/js/jquery.dataTables.min.js"></script><link href="Introduction_to_TMBleR_files/crosstalk-1.1.0.1/css/crosstalk.css" rel="stylesheet">
<script src="Introduction_to_TMBleR_files/crosstalk-1.1.0.1/js/crosstalk.min.js"></script><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>An introduction to TMBleR</h1>
            
            <h4 class="date">2021-06-21</h4>
      
      
      <div class="hidden name"><code>Introduction_to_TMBleR.Rmd</code></div>

    </div>

    
    
<script type="text/javascript">
document.addEventListener("DOMContentLoaded", function() {
  document.querySelector("h1").className = "title";
});
</script><script type="text/javascript">
document.addEventListener("DOMContentLoaded", function() {
  var links = document.links;  
  for (var i = 0, linksLength = links.length; i < linksLength; i++)
    if (links[i].hostname != window.location.hostname)
      links[i].target = '_blank';
});
</script><style>
body {
text-align: justify}
</style>
<div id="introduction" class="section level1">
<h1 class="hasAnchor">
<a href="#introduction" class="anchor"></a>Introduction</h1>
<p><em>Tumor mutational burden</em> (TMB), the total amount of somatic coding mutations in a tumor, represents a promising bio-marker to predict immunotherapy response in cancer patients. Initial studies investigating TMB used whole exome sequencing (WES) for quantification. Recently, it was demonstrated that TMB can be accurately measured also from targeted sequencing gene panels, which consistently reduces sequencing costs and paves the way for TMB routine adoption in clinical care. Currently, different sets of mutations are used to perform TMB quantification (e.g. only <em>non-synonymous mutations</em>, both <em>synonymous</em> and <em>non-synonymous mutations</em>, all but known cancer mutations as they are over-represented in gene panels) and no common standard exists. This hampers fair comparison of results from different studies and different gene panels. <strong>TMBleR allows to perform parallel panel-based TMB quantification by different combinations of variants filtering (e.g. synonymous, known cancer mutations, by VAF) and to easily plot estimated TMB values and visualize differences between methods and/or gene panels.</strong></p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va">TMBleR</span><span class="op">)</span></code></pre></div>
<p>This package finds to following important applications:</p>
<ul>
<li>
<a href="#partA">Calculate the TMB for cancer patients undergoing genomic sequencing;</a><a href="#fn1" class="footnote-ref" id="fnref1"><sup>1</sup></a>
</li>
<li><a href="#partB">Investigate how well the estimated TMB value from a genomic panel correlates to a whole exome sequencing (WES);</a></li>
<li><a href="#partC">Guide the design of a clinical genomics targeted panel by evaluate TMB performance to predict immunotherapy responders;</a></li>
</ul>
</div>
<div id="partA" class="section level1">
<h1 class="hasAnchor">
<a href="#partA" class="anchor"></a>TMB quantification using different filters</h1>
<p>In the example here provided, we want to perform TMB quantification of 2 different <a href="https://www.horizondiscovery.com">Horizon cell lines</a> sequenced by a targeted gene panel<a href="#fn2" class="footnote-ref" id="fnref2"><sup>2</sup></a>. Different filters can be applied on the variants reported in the corresponding vcf files, prior to TMB quantification. Indeed, panel-based TMB quantification requires a method to extrapolate the number of mutations observed in the narrow sequencing space of gene panels to the whole exome. For this reason, <em>synonymous mutations</em> can be counted for panel-based TMB quantification: although not immunogenic, they increase the statistical sampling size and reduce the noise. Besides panel size, gene composition also affects the accuracy of panel-based TMB quantification. In many gene panels, the over-representation of genes known to be highly mutated in cancers may lead to TMB overestimation. Therefore, some studies try to avoid this bias by removing somatic COSMIC variants or truncating mutations in tumor suppressors from variants used for TMB quantification. Other filters can be applied to variants for TMB quantification, such as minimum VAF thresholds. No common standard currently exists and we apply here different filters and compare the obtained TMB.</p>
<div class="figure">
<img src="workflow_tmbtool_p1.png" alt=""><p class="caption">Figure: TMBleR worflow schematic for simple TMB calculation</p>
</div>
<div id="dataset-preparation" class="section level3">
<h3 class="hasAnchor">
<a href="#dataset-preparation" class="anchor"></a>Dataset preparation</h3>
<p>There are 3 inputs required to perform the analysis:</p>
<ul>
<li>
<a href="#genome-assembly">Genome assemby</a> of reference: either “hg19” or “hg38”;</li>
<li>
<a href="#somatic-mutations-identified-by-the-gene-panel">Somatic mutations</a> identified by the gene panel, provided in vcf file format;</li>
<li>
<a href="#panel-design">Panel design</a>, provided as <code>.bed</code> file or <code>.txt</code> file with list of gene IDs.</li>
</ul>
<p>Finally, in order for the package to work at his full potential, optionally you might want to provide the latest COMISC somatic pathogenic SNVs. This is optional and allows to use the <code>remove.cancer</code> filter option of applyFilter(). This will be discussed further below:</p>
<ul>
<li>
<a href="#external-dataset-optional">External dataset</a> (Optional)</li>
</ul>
<div id="genome-assembly" class="section level4">
<h4 class="hasAnchor">
<a href="#genome-assembly" class="anchor"></a>Genome assembly</h4>
<p>Specify human assembly of reference, either “hg19” or “hg38”.</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">genome_assembly</span> <span class="op">&lt;-</span> <span class="st">"hg19"</span></code></pre></div>
</div>
<div id="somatic-mutations-identified-by-the-gene-panel" class="section level4">
<h4 class="hasAnchor">
<a href="#somatic-mutations-identified-by-the-gene-panel" class="anchor"></a>Somatic mutations identified by the gene panel</h4>
<p>Read vcf files containing somatic variants identified by panel sequencing.</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># Provide vcf_files as a list ---------------------------------</span>
<span class="co"># Note that even when reading only one file you </span>
<span class="co"># need to provide it as a list with element name corresponding to the name to </span>
<span class="co"># give to the sample and value corresponding to the path to the file </span>
<span class="co"># (i.e. vcf_file=list(LungSample="inst/extdata/LungSample.vcf"))</span>
<span class="va">vcf_files</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html">list</a></span><span class="op">(</span>Horizon5<span class="op">=</span><span class="st">"Horizon5_ExamplePanel.vcf"</span>, 
                  HorizonFFPEmild<span class="op">=</span><span class="st">"HorizonFFPEmild_ExamplePanel.vcf"</span><span class="op">)</span>
<span class="co"># For each vcf file, get the absolute path. ------------------</span>
<span class="co"># Here we are creating absolute paths by using .vcf files that were</span>
<span class="co"># bundled with the package </span>
<span class="va">vcf_files</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/lapply.html">lapply</a></span><span class="op">(</span><span class="va">vcf_files</span>, <span class="kw">function</span><span class="op">(</span><span class="va">x</span><span class="op">)</span> 
  <span class="fu"><a href="https://rdrr.io/r/base/system.file.html">system.file</a></span><span class="op">(</span><span class="st">"extdata"</span>, <span class="va">x</span>, package <span class="op">=</span> <span class="st">"TMBleR"</span>, mustWork <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span><span class="op">)</span>
<span class="co"># read in the files -------------------------------------------</span>
<span class="va">vcfs</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/readVcfFiles.html">readVcfFiles</a></span><span class="op">(</span>vcfFiles <span class="op">=</span> <span class="va">vcf_files</span>, assembly <span class="op">=</span> <span class="va">genome_assembly</span><span class="op">)</span></code></pre></div>
</div>
<div id="panel-design" class="section level4">
<h4 class="hasAnchor">
<a href="#panel-design" class="anchor"></a>Panel Design</h4>
<p>Read in input the panel sequencing design (either a <code>.bed</code> file with genomic coordinates of the targeted genes or <code>.txt</code> file with targeted gene names).</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># Get absolute path</span>
<span class="va">design_file_path</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html">system.file</a></span><span class="op">(</span> <span class="st">"extdata"</span>
                        , <span class="st">"ExamplePanel_GeneIDs.txt"</span>
                        , package <span class="op">=</span> <span class="st">"TMBleR"</span>
                        , mustWork <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span>

<span class="co"># Read bed file with the panel sequencing design </span>
<span class="va">design</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/readDesign.html">readDesign</a></span><span class="op">(</span>
  filename <span class="op">=</span> <span class="va">design_file_path</span>
  , assembly <span class="op">=</span> <span class="va">genome_assembly</span>
  , ids <span class="op">=</span> <span class="st">"entrezgene_id"</span><span class="op">)</span>
<span class="co">#&gt; Warning: `select_()` is deprecated as of dplyr 0.7.0.</span>
<span class="co">#&gt; Please use `select()` instead.</span>
<span class="co">#&gt; This warning is displayed once every 8 hours.</span>
<span class="co">#&gt; Call `lifecycle::last_warnings()` to see where this warning was generated.</span>
<span class="co">#&gt; Warning: `filter_()` is deprecated as of dplyr 0.7.0.</span>
<span class="co">#&gt; Please use `filter()` instead.</span>
<span class="co">#&gt; See vignette('programming') for more help</span>
<span class="co">#&gt; This warning is displayed once every 8 hours.</span>
<span class="co">#&gt; Call `lifecycle::last_warnings()` to see where this warning was generated.</span>
<span class="co">#&gt; Cache found</span>

<span class="co"># Preview</span>
<span class="fu"><a href="https://rdrr.io/r/utils/head.html">head</a></span><span class="op">(</span><span class="va">design</span><span class="op">)</span>
<span class="co">#&gt; GRanges object with 6 ranges and 1 metadata column:</span>
<span class="co">#&gt;       seqnames              ranges strand | ensembl_exon_id</span>
<span class="co">#&gt;          &lt;Rle&gt;           &lt;IRanges&gt;  &lt;Rle&gt; |     &lt;character&gt;</span>
<span class="co">#&gt;   [1]     chr1 244006427-244006538      - | ENSE00002329149</span>
<span class="co">#&gt;   [2]     chr1 243858893-243859018      - | ENSE00003573177</span>
<span class="co">#&gt;   [3]     chr1 243828074-243828185      - | ENSE00003478804</span>
<span class="co">#&gt;   [4]     chr1 243809195-243809339      - | ENSE00000815947</span>
<span class="co">#&gt;   [5]     chr1 243800913-243801044      - | ENSE00001069794</span>
<span class="co">#&gt;   [6]     chr1 243778398-243778463      - | ENSE00001358887</span>
<span class="co">#&gt;   -------</span>
<span class="co">#&gt;   seqinfo: 24 sequences from an unspecified genome; no seqlengths</span></code></pre></div>
</div>
<div id="external-dataset-optional" class="section level4">
<h4 class="hasAnchor">
<a href="#external-dataset-optional" class="anchor"></a>External dataset (Optional)</h4>
<p>External dataset are required for one of the mutation filters that removes known “cancer” mutation. This is achieved by setting <code>remove.cancer = TRUE</code> when calling <code>applyFilter()</code> (see section paragraph). Since these files are quite large, TMBleR comes with just a very small version that can be used to demo the package. If you are planning to use the <code>remove.cancer</code> filter, please see the vignette <strong>HOWTO_Import_external_data</strong> which contains the instructions on how to retrieve the COSMIC data and import it into TMBleR.</p>
</div>
</div>
<div id="mutation-filtering" class="section level3">
<h3 class="hasAnchor">
<a href="#mutation-filtering" class="anchor"></a>Mutation filtering</h3>
<p>Apply different filters on the identified variants. Typically, known cancer mutations are removed when performing panel-based TMB quantification, because the enrichment in cancer genes of these panels may lead to TMB overestimation. Synonymous mutations instead are removed when using WES datasets (in principle they do not generate tumor neoepitopes), whereas they are kept when using panels to increase statistical sampling size and reduce noise. Here we showcase how you can set up 7 different filtering combinations that will later on be used to estimate the TMB:</p>
<ul>
<li>Using an external dataset (downloaded from the Catalogue of Somatic Mutations in Cancer - COSMIC <a href="https://cancer.sanger.ac.uk/cosmic">https://cancer.sanger.ac.uk/cosmic</a>) to filter for pathogenic SNV mutations (e.g coding mutations described in COSMIC and truncating mutations in tumor suppressor genes). This is achieved by setting <code>remove.cancer = TRUE</code>;</li>
<li>filter one or a combination of mutations classes:
<ul>
<li>“synonymous”</li>
<li>“non-synonymous”</li>
<li>“frameshift”</li>
<li>“nonsense” (Usage example: <code>variantType = c("synonymous", "frameshift)</code>)</li>
</ul>
</li>
<li>Filer by Variant Allele Frequency (VAF), SNV with VAF lower than a specified threshold (defined by the <code>vaf.cutoff</code> argument) will be filtered out;</li>
<li>a combination of all the above.</li>
</ul>
<p>When setting <code>remove.cancer = T</code>, TMBleR will filter out cancer mutation from the <code>vcfs</code> object. If you did not provide a cancer dataset, a demo dataset will be loaded instead and a warning message will be print out.</p>
<p><strong>Note</strong> the results here shown use a under-sampled vcf from the COSMIC file. This data are only for illustration purposes.</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># Filter 1 ------------------------------------------------------------</span>
<span class="co"># Remove known cancer mutations using COSMIC data</span>
<span class="va">vcfs_NoCancer</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/applyFilters.html">applyFilters</a></span><span class="op">(</span>  vcfs <span class="op">=</span> <span class="va">vcfs</span>
                              , assembly <span class="op">=</span> <span class="va">genome_assembly</span>
                              , design <span class="op">=</span> <span class="va">design</span>
                              , remove.cancer <span class="op">=</span> <span class="cn">T</span><span class="op">)</span>
<span class="co">#&gt; Warning in filterCOSMIC(vcf_NoTruncatingInTS, assembly): data object COSMIC_hg19</span>
<span class="co">#&gt; was not found. Loading a COSMIC demo dataset. Please do not use this for TMB</span>
<span class="co">#&gt; analysis but only for demoing TMBleR</span>

<span class="co">#&gt; Warning in filterCOSMIC(vcf_NoTruncatingInTS, assembly): data object COSMIC_hg19</span>
<span class="co">#&gt; was not found. Loading a COSMIC demo dataset. Please do not use this for TMB</span>
<span class="co">#&gt; analysis but only for demoing TMBleR</span>

<span class="co"># Filter 2 ------------------------------------------------------------</span>
<span class="co"># Remove synonymous mutations</span>
<span class="va">vcfs_NoSynonymous</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/applyFilters.html">applyFilters</a></span><span class="op">(</span>vcfs <span class="op">=</span> <span class="va">vcfs</span>, 
                                  assembly <span class="op">=</span> <span class="va">genome_assembly</span>, 
                                  design <span class="op">=</span> <span class="va">design</span>, 
                                  variantType <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"synonymous"</span><span class="op">)</span><span class="op">)</span>

<span class="co"># Filter 3 ------------------------------------------------------------</span>
<span class="co"># Remove known cancer mutations and synonymous mutations</span>
<span class="va">vcfs_NoCancer_NoSynonymous</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/applyFilters.html">applyFilters</a></span><span class="op">(</span>vcfs <span class="op">=</span> <span class="va">vcfs</span>, 
                                           assembly <span class="op">=</span> <span class="va">genome_assembly</span>,
                                           design <span class="op">=</span> <span class="va">design</span>, 
                                           remove.cancer <span class="op">=</span> <span class="cn">T</span>, 
                                           variantType <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"synonymous"</span><span class="op">)</span><span class="op">)</span>
<span class="co">#&gt; Warning in filterCOSMIC(vcf_NoTruncatingInTS, assembly): data object COSMIC_hg19</span>
<span class="co">#&gt; was not found. Loading a COSMIC demo dataset. Please do not use this for TMB</span>
<span class="co">#&gt; analysis but only for demoing TMBleR</span>

<span class="co">#&gt; Warning in filterCOSMIC(vcf_NoTruncatingInTS, assembly): data object COSMIC_hg19</span>
<span class="co">#&gt; was not found. Loading a COSMIC demo dataset. Please do not use this for TMB</span>
<span class="co">#&gt; analysis but only for demoing TMBleR</span>

<span class="co"># Filter 4 ------------------------------------------------------------</span>
<span class="co"># Remove synonymous mutations and mutations with variant </span>
<span class="co"># allele frequency &lt; 0.05</span>
<span class="va">vcfs_NoSynonymous_VAFFilter</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/applyFilters.html">applyFilters</a></span><span class="op">(</span>vcfs <span class="op">=</span> <span class="va">vcfs</span>, 
                                            assembly <span class="op">=</span> <span class="va">genome_assembly</span>, 
                                            design <span class="op">=</span> <span class="va">design</span>, 
                                            vaf.cutoff <span class="op">=</span> <span class="fl">0.05</span>, 
                                            variantType <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"synonymous"</span><span class="op">)</span><span class="op">)</span>

<span class="co"># Filter 5 ------------------------------------------------------------</span>
<span class="co"># Remove mutations with variant allele frequency &lt; 0.05</span>
<span class="va">vcfs_VAFFilter</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/applyFilters.html">applyFilters</a></span><span class="op">(</span>
  vcfs <span class="op">=</span> <span class="va">vcfs</span>, 
  assembly <span class="op">=</span> <span class="va">genome_assembly</span>, 
  design <span class="op">=</span> <span class="va">design</span>, 
  vaf.cutoff <span class="op">=</span> <span class="fl">0.05</span>,
  tsList <span class="op">=</span> <span class="cn">NULL</span>, 
<span class="op">)</span>

<span class="co"># Filter 6 ------------------------------------------------------------</span>
<span class="co"># Remove known cancer mutations (e.g. coding mutations described in COSMIC and </span>
<span class="co"># truncating mutations in tumor suppressor genes) and mutations with variant </span>
<span class="co"># allele frequency &lt; 0.05</span>
<span class="va">vcfs_NoCancer_VAFFilter</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/applyFilters.html">applyFilters</a></span><span class="op">(</span>
  vcfs <span class="op">=</span> <span class="va">vcfs</span>, 
  assembly <span class="op">=</span> <span class="va">genome_assembly</span>, 
  design <span class="op">=</span> <span class="va">design</span>, 
  vaf.cutoff <span class="op">=</span> <span class="fl">0.05</span>, 
  remove.cancer <span class="op">=</span> <span class="cn">T</span>, 
<span class="op">)</span>
<span class="co">#&gt; Warning in filterCOSMIC(vcf_NoTruncatingInTS, assembly): data object COSMIC_hg19</span>
<span class="co">#&gt; was not found. Loading a COSMIC demo dataset. Please do not use this for TMB</span>
<span class="co">#&gt; analysis but only for demoing TMBleR</span>

<span class="co">#&gt; Warning in filterCOSMIC(vcf_NoTruncatingInTS, assembly): data object COSMIC_hg19</span>
<span class="co">#&gt; was not found. Loading a COSMIC demo dataset. Please do not use this for TMB</span>
<span class="co">#&gt; analysis but only for demoing TMBleR</span>

<span class="co"># Filter 7 ------------------------------------------------------------</span>
<span class="co"># Remove known cancer mutations (e.g. coding mutations described in COSMIC and </span>
<span class="co"># truncating mutations in tumor suppressor genes), synonymous mutations and </span>
<span class="co"># mutations with variant allele frequency &lt; 0.05</span>
<span class="va">vcfs_NoCancer_VAFFilter_NoSynonymous</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/applyFilters.html">applyFilters</a></span><span class="op">(</span>
  vcfs <span class="op">=</span> <span class="va">vcfs</span>, 
  assembly <span class="op">=</span> <span class="va">genome_assembly</span>, 
  design <span class="op">=</span> <span class="va">design</span>, 
  vaf.cutoff <span class="op">=</span> <span class="fl">0.05</span>, 
  remove.cancer <span class="op">=</span> <span class="cn">T</span>, 
  variantType <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"synonymous"</span><span class="op">)</span>
<span class="op">)</span>
<span class="co">#&gt; Warning in filterCOSMIC(vcf_NoTruncatingInTS, assembly): data object COSMIC_hg19</span>
<span class="co">#&gt; was not found. Loading a COSMIC demo dataset. Please do not use this for TMB</span>
<span class="co">#&gt; analysis but only for demoing TMBleR</span>

<span class="co">#&gt; Warning in filterCOSMIC(vcf_NoTruncatingInTS, assembly): data object COSMIC_hg19</span>
<span class="co">#&gt; was not found. Loading a COSMIC demo dataset. Please do not use this for TMB</span>
<span class="co">#&gt; analysis but only for demoing TMBleR</span>

<span class="co"># We also want to perform TMB quantification on a non filtered vcf. Therefore </span>
<span class="co"># we need to format it as required by the applyTMB function. We can </span>
<span class="co"># do so by simply providing no Filtering options</span>
<span class="va">vcfs_nonfiltered</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/applyFilters.html">applyFilters</a></span><span class="op">(</span>
  vcfs <span class="op">=</span> <span class="va">vcfs</span>,
  assembly <span class="op">=</span> <span class="va">genome_assembly</span>, 
  design <span class="op">=</span> <span class="va">design</span>, 
<span class="op">)</span></code></pre></div>
</div>
<div id="tmb-quantification" class="section level3">
<h3 class="hasAnchor">
<a href="#tmb-quantification" class="anchor"></a>TMB quantification</h3>
<p>The following command allows to estimate the TMB from the different filtering criteria chosen.</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># Concatenate all filtered and unfiltered vcfs</span>
<span class="va">vcfs_all</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="va">vcfs_NoCancer</span>, 
              <span class="va">vcfs_NoCancer_NoSynonymous</span>, 
              <span class="va">vcfs_NoSynonymous</span>, 
              <span class="va">vcfs_NoSynonymous_VAFFilter</span>, 
              <span class="va">vcfs_VAFFilter</span>, 
              <span class="va">vcfs_NoCancer_VAFFilter</span>, 
              <span class="va">vcfs_NoCancer_VAFFilter_NoSynonymous</span>, 
              <span class="va">vcfs_nonfiltered</span> <span class="op">)</span>

<span class="co"># Perform TMB quantification</span>
<span class="va">TMB_res</span><span class="op">=</span><span class="fu"><a href="../reference/applyTMB.html">applyTMB</a></span><span class="op">(</span>inputForTMB <span class="op">=</span> <span class="va">vcfs_all</span>, assembly <span class="op">=</span> <span class="va">genome_assembly</span><span class="op">)</span>

<span class="co"># Print results in table format</span>
<span class="fu">DT</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/DT/man/datatable.html">datatable</a></span><span class="op">(</span><span class="va">TMB_res</span><span class="op">)</span></code></pre></div>
<div id="htmlwidget-68abc20c5d0ed70902fa" style="width:100%;height:auto;" class="datatables html-widget"></div>
<script type="application/json" data-for="htmlwidget-68abc20c5d0ed70902fa">{"x":{"filter":"none","data":[["1","2","3","4","5","6","7","8","9","10","11","12","13","14","15","16"],["Horizon5","HorizonFFPEmild","Horizon5","HorizonFFPEmild","Horizon5","HorizonFFPEmild","Horizon5","HorizonFFPEmild","Horizon5","HorizonFFPEmild","Horizon5","HorizonFFPEmild","Horizon5","HorizonFFPEmild","Horizon5","HorizonFFPEmild"],["ExamplePanel_GeneIDs","ExamplePanel_GeneIDs","ExamplePanel_GeneIDs","ExamplePanel_GeneIDs","ExamplePanel_GeneIDs","ExamplePanel_GeneIDs","ExamplePanel_GeneIDs","ExamplePanel_GeneIDs","ExamplePanel_GeneIDs","ExamplePanel_GeneIDs","ExamplePanel_GeneIDs","ExamplePanel_GeneIDs","ExamplePanel_GeneIDs","ExamplePanel_GeneIDs","ExamplePanel_GeneIDs","ExamplePanel_GeneIDs"],["vaf=0NoCancerMuts","vaf=0NoCancerMuts","vaf=0NoCancerMutsNosynonymous","vaf=0NoCancerMutsNosynonymous","vaf=0Nosynonymous","vaf=0Nosynonymous","vaf=0.05Nosynonymous","vaf=0.05Nosynonymous","vaf=0.05","vaf=0.05","vaf=0.05NoCancerMuts","vaf=0.05NoCancerMuts","vaf=0.05NoCancerMutsNosynonymous","vaf=0.05NoCancerMutsNosynonymous","vaf=0","vaf=0"],[1.861301,1.861301,1.861301,1.861301,1.861301,1.861301,1.861301,1.861301,1.861301,1.861301,1.861301,1.861301,1.861301,1.861301,1.861301,1.861301],[189,182,98,97,104,105,45,84,78,164,73,149,41,76,203,198],[101.54,97.78,52.65,52.11,55.87,56.41,24.18,45.13,41.91,88.11,39.22,80.05,22.03,40.83,109.06,106.38]],"container":"<table class=\"display\">\n  <thead>\n    <tr>\n      <th> <\/th>\n      <th>Sample<\/th>\n      <th>Design<\/th>\n      <th>Filter<\/th>\n      <th>Sequencing_Size<\/th>\n      <th>Tot_Number_Mutations<\/th>\n      <th>TMB_per_Mb<\/th>\n    <\/tr>\n  <\/thead>\n<\/table>","options":{"columnDefs":[{"className":"dt-right","targets":[4,5,6]},{"orderable":false,"targets":0}],"order":[],"autoWidth":false,"orderClasses":false}},"evals":[],"jsHooks":[]}</script><p>The TMB estimate results can be preview in different ways:</p>
<ul>
<li>Classical barplot</li>
</ul>
<p>where the samples are grouped together on the x-axis, while the TMB value per Mb is shown on the y-axis. Each filter is displayed with a different color.</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># Plot results using barplots</span>
<span class="fu"><a href="../reference/plotTMB.html">plotTMB</a></span><span class="op">(</span>TMB_results <span class="op">=</span> <span class="va">TMB_res</span>, type<span class="op">=</span><span class="st">"barplot"</span><span class="op">)</span></code></pre></div>
<p><img src="Introduction_to_TMBleR_files/figure-html/unnamed-chunk-8-1.png" width="700"></p>
<p>When processing a large number of samples, the previous plot might be a little bit cluttered. Here we present an alternative visualization that groups together the different samples showing their TMB value density for each filter.</p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># Plot results using density plots</span>
<span class="fu"><a href="../reference/plotTMB.html">plotTMB</a></span><span class="op">(</span>TMB_results <span class="op">=</span> <span class="va">TMB_res</span>, type<span class="op">=</span><span class="st">"densityplot"</span><span class="op">)</span></code></pre></div>
<p><img src="Introduction_to_TMBleR_files/figure-html/unnamed-chunk-9-1.png" width="768"></p>
<p>Finally, a boxplot can also provide additional information on the distribution</p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># Plot results with boxplot</span>
<span class="fu"><a href="../reference/plotTMB.html">plotTMB</a></span><span class="op">(</span>TMB_results <span class="op">=</span> <span class="va">TMB_res</span>, type<span class="op">=</span><span class="st">"boxplot"</span><span class="op">)</span></code></pre></div>
<p><img src="Introduction_to_TMBleR_files/figure-html/unnamed-chunk-10-1.png" width="700"></p>
</div>
</div>
<div id="partB" class="section level1">
<h1 class="hasAnchor">
<a href="#partB" class="anchor"></a>Simulate TMB quantification from a targeted gene panel</h1>
<p>In case we are investigating a targeted gene panel design and we are interest in its performance to evaluate TMB but no sequencing data is available, then we use WES dataset to simulate it. This section covers an example on how to just do that; we will subset a WES dataset to only include the genes targeted by the panel in examination. Then we compare the accuracy of the (simulated) panel-based TMB with the reference value of WES-based TMB. This is a common strategy for in-silico validation of accuracy and predictive power of panel-based TMB.</p>
<div id="dataset-preparation-1" class="section level3">
<h3 class="hasAnchor">
<a href="#dataset-preparation-1" class="anchor"></a>Dataset preparation</h3>
<p>There are 4 inputs required to perform the analysis:</p>
<ul>
<li>
<a href="#genome-assembly">Genome assemby</a> of reference: either “hg19” or “hg38”;</li>
<li>
<a href="#somatic-mutations-identified-by-wes">Somatic mutations identified by WES</a>;</li>
<li>
<a href="#wes-and-gene-panel-design">Design of WES</a>, provided as <code>.bed</code> file or <code>.txt</code> file with list of gene IDs;</li>
<li>
<a href="#wes-and-gene-panel-design">Design of the panel to simulate</a>, provided as BED file or TXT file with list of gene IDs.</li>
</ul>
<div id="genome-assembly-1" class="section level4">
<h4 class="hasAnchor">
<a href="#genome-assembly-1" class="anchor"></a>Genome assembly</h4>
<p>Specify human assembly of reference, either “hg19” or “hg38”</p>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># 1 . =========================================================================</span>
<span class="va">genome_assembly</span> <span class="op">&lt;-</span> <span class="st">"hg19"</span></code></pre></div>
</div>
<div id="somatic-mutations-identified-by-wes" class="section level4">
<h4 class="hasAnchor">
<a href="#somatic-mutations-identified-by-wes" class="anchor"></a>Somatic mutations identified by WES</h4>
<p>The mutational profile is to be provided as a <code>.vcf</code> file. For this example we have already prepared a showcase dataset.</p>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># Load built-in list of 4 vcfs generated by WES</span>
<span class="co"># 2. ==========================================================================</span>
<span class="co"># Read file names</span>
<span class="va">vcf_files</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html">list</a></span><span class="op">(</span>  Pat02<span class="op">=</span><span class="st">"Pat02_vs_NORMAL.vcf"</span>
                  , Pat03<span class="op">=</span><span class="st">"Pat03_vs_NORMAL.vcf"</span>
                  , Pat04<span class="op">=</span><span class="st">"Pat04_vs_NORMAL.vcf"</span>
                  , Pat06<span class="op">=</span><span class="st">"Pat06_vs_NORMAL.vcf"</span><span class="op">)</span>

<span class="co"># Find the absolute path of the vcf files bundled with the TMBleR package</span>
<span class="va">vcf_files</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/lapply.html">lapply</a></span><span class="op">(</span>
  <span class="va">vcf_files</span>
  , <span class="kw">function</span><span class="op">(</span><span class="va">x</span><span class="op">)</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html">system.file</a></span><span class="op">(</span><span class="st">"extdata"</span>, <span class="va">x</span>, package <span class="op">=</span> <span class="st">"TMBleR"</span>, mustWork <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span>
<span class="op">)</span>

<span class="co"># Read vcf files and put in a list</span>
<span class="va">ExampleWESvcfs</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/readVcfFiles.html">readVcfFiles</a></span><span class="op">(</span>vcfFiles<span class="op">=</span><span class="va">vcf_files</span>, <span class="va">genome_assembly</span><span class="op">)</span></code></pre></div>
</div>
<div id="wes-and-gene-panel-design" class="section level4">
<h4 class="hasAnchor">
<a href="#wes-and-gene-panel-design" class="anchor"></a>WES and gene panel design</h4>
<p>Import both the WES genomic coordinates as well as the panel design targeted regions. In this example we will be using 2 gene panel designs :</p>
<ul>
<li><code>ExamplePaneldesign.bed</code></li>
<li><code>MskImpactPanel_468GenesMutsAndCNA.txt</code></li>
</ul>
<p>Finally the WES genomic coordinates can be found here:</p>
<ul>
<li><code>SureSelect_Human_All_Exon_V2_Regions.bed</code></li>
</ul>
<p>(note: use the following command to preview the file <code><a href="https://rdrr.io/r/base/system.file.html">system.file("extdata","SureSelect_Human_All_Exon_V2_Regions.bed",package = "TMBleR",mustWork = TRUE)</a></code>).</p>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># Read files with WES and gene panel design</span>
<span class="co"># 3. ==========================================================================</span>

<span class="co"># First panel ------------------------------------------------</span>
<span class="co"># Load the design of the first gene panels that you want to simulate</span>
<span class="va">design_file_path</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html">system.file</a></span><span class="op">(</span><span class="st">"extdata"</span>,
                          <span class="st">"ExamplePaneldesign.bed"</span>,
                          package <span class="op">=</span> <span class="st">"TMBleR"</span>,
                          mustWork <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span>
<span class="va">ExamplePaneldesign</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/readDesign.html">readDesign</a></span><span class="op">(</span>filename <span class="op">=</span> <span class="va">design_file_path</span>
                                 , assembly <span class="op">=</span> <span class="st">"hg19"</span>
                                 , name <span class="op">=</span> <span class="st">"SimulatedPanel"</span><span class="op">)</span>


<span class="co"># Second Panel ------------------------------------------------</span>
<span class="va">design_file_path</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html">system.file</a></span><span class="op">(</span><span class="st">"extdata"</span>,
                         <span class="st">"MskImpactPanel_468GenesMutsAndCNA.txt"</span>,
                          package <span class="op">=</span> <span class="st">"TMBleR"</span>,
                          mustWork <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span>
<span class="va">MSKdesign</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/readDesign.html">readDesign</a></span><span class="op">(</span>filename <span class="op">=</span> <span class="va">design_file_path</span>
                        , assembly <span class="op">=</span> <span class="st">"hg19"</span>
                        , ids <span class="op">=</span> <span class="st">"hgnc_symbol"</span>
                        , name <span class="op">=</span> <span class="st">"MSKPanel"</span><span class="op">)</span>
<span class="co">#&gt; Cache found</span>


<span class="co"># Load the design of the WES ---------------------------------</span>
<span class="va">design_file_path</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html">system.file</a></span><span class="op">(</span><span class="st">"extdata"</span>,
                          <span class="st">"SureSelect_Human_All_Exon_V2_Regions.bed"</span>,
                          package <span class="op">=</span> <span class="st">"TMBleR"</span>,
                          mustWork <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span>
<span class="va">ExampleWESdesign</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/readDesign.html">readDesign</a></span><span class="op">(</span>filename <span class="op">=</span> <span class="va">design_file_path</span>
                               , assembly <span class="op">=</span> <span class="st">"hg19"</span>
                               , name <span class="op">=</span> <span class="st">"WES"</span><span class="op">)</span></code></pre></div>
</div>
</div>
<div id="mutation-filtering-1" class="section level3">
<h3 class="hasAnchor">
<a href="#mutation-filtering-1" class="anchor"></a>Mutation filtering</h3>
<p>In the TMB quantification of genomic panel, cancer mutations are usually removed to avoid overestimation of TMB due to panel enrichment in cancer genes. This can be achieved by filtering out known cancer mutations or truncating mutations in tumor suppressors ( <em>e.g.</em> <code>remove.cancer=T</code>). In the following example, we will apply 3 different TMB filters to the WES somatic dataset:</p>
<ul>
<li>
<strong>Filter 1</strong>: remove cancer mutation (as in F1CDx and TSO500);</li>
<li>
<strong>Filter 2</strong>: remove known cancer mutation AND any variant classified as either “synonymous”,“frameshift”, “nonsense” (as in MSK-IMPACT panel);</li>
<li>
<strong>Filter 3</strong>: as usually done in WES studies to estimate TMB, filter out only synonymous mutation.</li>
</ul>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># Filters for gene panel simulation ===================================</span>
<span class="co"># Filter 1 ------------------------------------------------------------</span>
<span class="co"># Remove known cancer mutations (e.g. coding mutations described in COSMIC </span>
<span class="co"># and truncating mutations in tumor suppressor genes)</span>
<span class="va">vcfs_NoCancer</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/applyFilters.html">applyFilters</a></span><span class="op">(</span>
  vcfs <span class="op">=</span> <span class="va">ExampleWESvcfs</span>
  , assembly <span class="op">=</span> <span class="va">genome_assembly</span>
  , design <span class="op">=</span> <span class="va">ExampleWESdesign</span>
  , remove.cancer <span class="op">=</span> <span class="cn">T</span>
<span class="op">)</span>

<span class="co"># Filter 2 ------------------------------------------------------------</span>
<span class="co"># Remove synonymous, indels and nonsense mutations </span>
<span class="va">vcfs_NoSynonymousNoFrameshiftNoNonsense</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/applyFilters.html">applyFilters</a></span><span class="op">(</span>
  vcfs <span class="op">=</span> <span class="va">ExampleWESvcfs</span>, 
  assembly <span class="op">=</span> <span class="va">genome_assembly</span>, 
  design <span class="op">=</span> <span class="va">ExampleWESdesign</span>, 
  variantType <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"synonymous"</span>, <span class="st">"frameshift"</span>, <span class="st">"nonsense"</span><span class="op">)</span><span class="op">)</span>

<span class="co"># Filter 3  ----------------------------------------------------------</span>
<span class="co"># Filter out synonymous mutations, as usually done for WES-based TMB </span>
<span class="co"># quantification. This filtered WES dataset will be used for WES-based TMB </span>
<span class="co"># quantification</span>
<span class="va">vcfs_NoSynonymous_WES</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/applyFilters.html">applyFilters</a></span><span class="op">(</span>
  vcfs <span class="op">=</span> <span class="va">ExampleWESvcfs</span>, 
  assembly <span class="op">=</span> <span class="va">genome_assembly</span>, 
  design <span class="op">=</span> <span class="va">ExampleWESdesign</span>, 
  variantType <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"synonymous"</span><span class="op">)</span><span class="op">)</span></code></pre></div>
</div>
<div id="subset-the-wes" class="section level3">
<h3 class="hasAnchor">
<a href="#subset-the-wes" class="anchor"></a>Subset the WES</h3>
<p>Now that we have applied different mutational filters on the somatic mutation in our WES dataset, we will try to simulate the panel by removing from the dataset the mutational profile that was “mapped” outside of the genomic panel of interest.</p>
<p>We will performed this process on the two panels we are currently investigating.</p>
<div class="sourceCode" id="cb14"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># PANEL 1 - generic example panel</span>
<span class="co"># =====================================================================</span>
<span class="co"># Subset WES with Filter 1</span>
<span class="va">SimulatedExamplePanel_NoCancer</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/applySimulatePanel.html">applySimulatePanel</a></span><span class="op">(</span>
  WES <span class="op">=</span> <span class="va">vcfs_NoCancer</span>,               <span class="co"># &lt;--- results from Filter 1</span>
  WES.design <span class="op">=</span> <span class="va">ExampleWESdesign</span>,     <span class="co"># &lt;--- This is the WES design </span>
  panel.design <span class="op">=</span> <span class="va">ExamplePaneldesign</span>, <span class="co"># &lt;--- this is design of the first panels</span>
  assembly <span class="op">=</span> <span class="va">genome_assembly</span>
<span class="op">)</span>
<span class="co"># Subset WES with Filter 2</span>
<span class="va">SimulatedExamplePanel_NoSynonymousNoFrameshiftNoNonsense</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/applySimulatePanel.html">applySimulatePanel</a></span><span class="op">(</span>
  WES <span class="op">=</span> <span class="va">vcfs_NoSynonymousNoFrameshiftNoNonsense</span>,  <span class="co"># &lt;--- results from Filter 2</span>
  WES.design <span class="op">=</span> <span class="va">ExampleWESdesign</span>,
  panel.design <span class="op">=</span> <span class="va">ExamplePaneldesign</span>, 
  assembly <span class="op">=</span> <span class="va">genome_assembly</span>
<span class="op">)</span>
<span class="co"># Concatenate the 2 resulting panels with different vcf filters</span>
<span class="va">vcfs_SimulatedExamplePanel</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span>
  <span class="va">SimulatedExamplePanel_NoCancer</span>
  , <span class="va">SimulatedExamplePanel_NoSynonymousNoFrameshiftNoNonsense</span>
<span class="op">)</span>


<span class="co"># Panel 2 - MSK panel</span>
<span class="co"># =====================================================================</span>
<span class="co"># Subset WES with Filter 1</span>
<span class="va">SimulatedMSKPanel_NoCancer</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/applySimulatePanel.html">applySimulatePanel</a></span><span class="op">(</span>
  WES <span class="op">=</span> <span class="va">vcfs_NoCancer</span>,            <span class="co"># &lt;--- results from Filter 1</span>
  WES.design <span class="op">=</span> <span class="va">ExampleWESdesign</span>,  <span class="co"># &lt;--- this is the WES design</span>
  panel.design <span class="op">=</span> <span class="va">MSKdesign</span>,       <span class="co"># &lt;--- this is the design of the second panel</span>
  assembly <span class="op">=</span> <span class="va">genome_assembly</span>
<span class="op">)</span>
<span class="co"># Subset WES with Filter 2</span>
<span class="va">SimulatedMSKPanel_NoSynonymousNoFrameshiftNoNonsense</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/applySimulatePanel.html">applySimulatePanel</a></span><span class="op">(</span>
  WES <span class="op">=</span> <span class="va">vcfs_NoSynonymousNoFrameshiftNoNonsense</span>,  <span class="co"># &lt;--- results from Filter 2</span>
  WES.design <span class="op">=</span> <span class="va">ExampleWESdesign</span>,
  panel.design <span class="op">=</span> <span class="va">MSKdesign</span>, 
  assembly <span class="op">=</span> <span class="va">genome_assembly</span>
<span class="op">)</span>
<span class="va">vcfs_SimulatedMSKPanel</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span>
  <span class="va">SimulatedMSKPanel_NoCancer</span>,
  <span class="va">SimulatedMSKPanel_NoSynonymousNoFrameshiftNoNonsense</span>
<span class="op">)</span></code></pre></div>
</div>
<div id="tmb-quantification-1" class="section level3">
<h3 class="hasAnchor">
<a href="#tmb-quantification-1" class="anchor"></a>TMB quantification</h3>
<p>After the previous step, the subset dataset will only contain variants in the regions targeted by the panel you want to simulate. We can now finally estimate the TMB. We will estimate the TMB for the following 3 scenarios:</p>
<ul>
<li>Simulated Example panel</li>
<li>Simulated MSK Panel</li>
<li>Original WES dataset</li>
</ul>
<p>We have included also the “original WES” dataset so that we can perform a comparison.</p>
<div class="sourceCode" id="cb15"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># TMB quantification from the simulated panels </span>
<span class="co"># ====================================================================</span>

<span class="co">#  Panel 1  ----------------------------------------------------------</span>
<span class="va">TMBs_SimulatedExamplePanel</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/applyTMB.html">applyTMB</a></span><span class="op">(</span>
  inputForTMB <span class="op">=</span> <span class="va">vcfs_SimulatedExamplePanel</span>
  , assembly <span class="op">=</span> <span class="va">genome_assembly</span>
<span class="op">)</span>
<span class="co"># generate plot</span>
<span class="fu"><a href="../reference/plotTMB.html">plotTMB</a></span><span class="op">(</span>TMB_results <span class="op">=</span> <span class="va">TMBs_SimulatedExamplePanel</span>, type<span class="op">=</span><span class="st">"barplot"</span><span class="op">)</span></code></pre></div>
<p><img src="Introduction_to_TMBleR_files/figure-html/unnamed-chunk-16-1.png" width="768"></p>
<div class="sourceCode" id="cb16"><pre class="downlit sourceCode r">
<code class="sourceCode R">

<span class="co">#  Panel 2  ----------------------------------------------------------</span>
<span class="va">TMBs_SimulatedMSKPanel</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/applyTMB.html">applyTMB</a></span><span class="op">(</span>
  inputForTMB <span class="op">=</span> <span class="va">vcfs_SimulatedMSKPanel</span>
  , assembly <span class="op">=</span> <span class="va">genome_assembly</span>
<span class="op">)</span>
<span class="co"># generate plot</span>
<span class="fu"><a href="../reference/plotTMB.html">plotTMB</a></span><span class="op">(</span>TMB_results <span class="op">=</span> <span class="va">TMBs_SimulatedMSKPanel</span>, type<span class="op">=</span><span class="st">"barplot"</span><span class="op">)</span></code></pre></div>
<p><img src="Introduction_to_TMBleR_files/figure-html/unnamed-chunk-16-2.png" width="768"></p>
<div class="sourceCode" id="cb17"><pre class="downlit sourceCode r">
<code class="sourceCode R">
<span class="co"># Original WES --------------------------------------------</span>
<span class="va">TMBs_WES</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/applyTMB.html">applyTMB</a></span><span class="op">(</span>inputForTMB <span class="op">=</span> <span class="va">vcfs_NoSynonymous_WES</span>
                     , assembly <span class="op">=</span> <span class="va">genome_assembly</span><span class="op">)</span>
<span class="co"># generate plot</span>
<span class="fu"><a href="../reference/plotTMB.html">plotTMB</a></span><span class="op">(</span>TMB_results <span class="op">=</span> <span class="va">TMBs_WES</span>, type<span class="op">=</span><span class="st">"barplot"</span><span class="op">)</span></code></pre></div>
<p><img src="Introduction_to_TMBleR_files/figure-html/unnamed-chunk-16-3.png" width="768"></p>
<p>The following code is just to merge together all the results from the 3 previous experiments into a single dataframe. We will have a new colum “Panel” that encodes the experiment</p>
<div class="sourceCode" id="cb18"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="http://purrr.tidyverse.org">purrr</a></span><span class="op">)</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; Attaching package: 'purrr'</span>
<span class="co">#&gt; The following object is masked from 'package:TMBleR':</span>
<span class="co">#&gt; </span>
<span class="co">#&gt;     set_names</span>
<span class="co">#&gt; The following object is masked from 'package:testthat':</span>
<span class="co">#&gt; </span>
<span class="co">#&gt;     is_null</span>
<span class="co"># Create a dataframe with the TMB results from different </span>
<span class="co"># panels  together all together</span>
<span class="va">TMB_res_df</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html">list</a></span><span class="op">(</span>SimulatedExamplePanel<span class="op">=</span><span class="va">TMBs_SimulatedExamplePanel</span>
                  , SimulatedMSKPanel<span class="op">=</span><span class="va">TMBs_SimulatedMSKPanel</span>
                  , WES<span class="op">=</span><span class="va">TMBs_WES</span><span class="op">)</span>  <span class="op">%&gt;%</span>
  <span class="co"># convert Factors into strings in the process. </span>
  <span class="fu">purrr</span><span class="fu">::</span><span class="fu"><a href="https://purrr.tidyverse.org/reference/map.html">map</a></span><span class="op">(</span> <span class="op">~</span> <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate_all.html">mutate_at</a></span><span class="op">(</span><span class="va">.</span>, <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"Sample"</span>, <span class="st">"Filter"</span>, <span class="st">"Design"</span><span class="op">)</span>, <span class="va">as.character</span><span class="op">)</span><span class="op">)</span> <span class="op">%&gt;%</span> 
  <span class="co"># make list into a dataframe</span>
  <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/bind.html">bind_rows</a></span><span class="op">(</span><span class="op">)</span>

<span class="co"># Preview</span>
<span class="fu">DT</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/DT/man/datatable.html">datatable</a></span><span class="op">(</span><span class="va">TMB_res_df</span><span class="op">)</span></code></pre></div>
<div id="htmlwidget-546677a3b83b28ec6884" style="width:100%;height:auto;" class="datatables html-widget"></div>
<script type="application/json" data-for="htmlwidget-546677a3b83b28ec6884">{"x":{"filter":"none","data":[["1","2","3","4","5","6","7","8","9","10","11","12","13","14","15","16","17","18","19","20"],["Pat02","Pat03","Pat04","Pat06","Pat02","Pat03","Pat04","Pat06","Pat02","Pat03","Pat04","Pat06","Pat02","Pat03","Pat04","Pat06","Pat02","Pat03","Pat04","Pat06"],["SimulatedPanel","SimulatedPanel","SimulatedPanel","SimulatedPanel","SimulatedPanel","SimulatedPanel","SimulatedPanel","SimulatedPanel","MSKPanel","MSKPanel","MSKPanel","MSKPanel","MSKPanel","MSKPanel","MSKPanel","MSKPanel","WES","WES","WES","WES"],["vaf=0NoCancerMuts","vaf=0NoCancerMuts","vaf=0NoCancerMuts","vaf=0NoCancerMuts","vaf=0NosynonymousNoframeshiftNononsense","vaf=0NosynonymousNoframeshiftNononsense","vaf=0NosynonymousNoframeshiftNononsense","vaf=0NosynonymousNoframeshiftNononsense","vaf=0NoCancerMuts","vaf=0NoCancerMuts","vaf=0NoCancerMuts","vaf=0NoCancerMuts","vaf=0NosynonymousNoframeshiftNononsense","vaf=0NosynonymousNoframeshiftNononsense","vaf=0NosynonymousNoframeshiftNononsense","vaf=0NosynonymousNoframeshiftNononsense","vaf=0Nosynonymous","vaf=0Nosynonymous","vaf=0Nosynonymous","vaf=0Nosynonymous"],[2.727206,2.727206,2.727206,2.727206,2.727206,2.727206,2.727206,2.727206,3.066137,3.066137,3.066137,3.066137,3.066137,3.066137,3.066137,3.066137,46.401091,46.401091,46.401091,46.401091],[18,29,26,12,15,19,21,9,9,19,12,7,9,12,11,5,203,332,316,136],[6.6,10.63,9.53,4.4,5.5,6.97,7.7,3.3,2.94,6.2,3.91,2.28,2.94,3.91,3.59,1.63,4.37,7.16,6.81,2.93]],"container":"<table class=\"display\">\n  <thead>\n    <tr>\n      <th> <\/th>\n      <th>Sample<\/th>\n      <th>Design<\/th>\n      <th>Filter<\/th>\n      <th>Sequencing_Size<\/th>\n      <th>Tot_Number_Mutations<\/th>\n      <th>TMB_per_Mb<\/th>\n    <\/tr>\n  <\/thead>\n<\/table>","options":{"columnDefs":[{"className":"dt-right","targets":[4,5,6]},{"orderable":false,"targets":0}],"order":[],"autoWidth":false,"orderClasses":false}},"evals":[],"jsHooks":[]}</script><p>The same bar plot as before can now be printed all together into one single plot calling the very same fuction.</p>
<div class="sourceCode" id="cb19"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="../reference/plotTMB.html">plotTMB</a></span><span class="op">(</span><span class="va">TMB_res_df</span><span class="op">)</span></code></pre></div>
<p><img src="Introduction_to_TMBleR_files/figure-html/unnamed-chunk-18-1.png" width="700"></p>
</div>
<div id="plot-tmb-distribution-by-groups" class="section level3">
<h3 class="hasAnchor">
<a href="#plot-tmb-distribution-by-groups" class="anchor"></a>Plot TMB distribution by groups</h3>
<p>To simplify interpretation, the results can be grouped in</p>
<ul>
<li>by filters</li>
<li>by design (e.g. panel or WES)</li>
</ul>
<div class="sourceCode" id="cb20"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="../reference/plotTMB.html">plotTMB</a></span><span class="op">(</span><span class="va">TMB_res_df</span>, type <span class="op">=</span><span class="st">"densityplot"</span><span class="op">)</span></code></pre></div>
<p><img src="Introduction_to_TMBleR_files/figure-html/unnamed-chunk-19-1.png" width="768"></p>
<p>For a more compact plot, we can use boxplots</p>
<div class="sourceCode" id="cb21"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="../reference/plotTMB.html">plotTMB</a></span><span class="op">(</span><span class="va">TMB_res_df</span>, type<span class="op">=</span><span class="st">"boxplot"</span><span class="op">)</span></code></pre></div>
<p><img src="Introduction_to_TMBleR_files/figure-html/unnamed-chunk-20-1.png" width="768"></p>
</div>
<div id="compare-simulated-panel-based-tmb-to-the-reference-wes-based-tmb-values" class="section level3">
<h3 class="hasAnchor">
<a href="#compare-simulated-panel-based-tmb-to-the-reference-wes-based-tmb-values" class="anchor"></a>Compare simulated panel-based TMB to the reference WES-based TMB values</h3>
<p>Here we showcase how you can use the function <code><a href="../reference/correlateTMBvalues.html">correlateTMBvalues()</a></code> to run a quick correlation analysis on two variable: the TMB generated by one of the filters and a comparison TMB from the original WES.</p>
<p>In the following example we will use Spearman correlation coefficient between the <strong>ExamplePanel</strong> and the original WES dataset both filtered out by removing the pathogenic mutations only (i.e. <code>noCancerMuts</code>, the other filters are not considered in this analysis).</p>
<div class="sourceCode" id="cb22"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># Spearman correlation -------------------------------------------------------</span>
<span class="co"># Plot Spearman correlation</span>
<span class="co">#par(cex.axis=1.2, cex.lab=1.2)</span>
<span class="va">TMBs_SimulatedExamplePanel</span> <span class="op">%&gt;%</span>
  <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html">filter</a></span><span class="op">(</span><span class="va">Filter</span> <span class="op">==</span> <span class="st">"vaf=0NoCancerMuts"</span><span class="op">)</span> <span class="op">%&gt;%</span>
  <span class="va">.</span><span class="op">$</span><span class="va">TMB_per_Mb</span> <span class="op">%&gt;%</span>
  <span class="fu"><a href="../reference/correlateTMBvalues.html">correlateTMBvalues</a></span><span class="op">(</span>panel.TMB <span class="op">=</span> <span class="va">.</span>,
                     WES.TMB <span class="op">=</span> <span class="va">TMBs_WES</span><span class="op">$</span><span class="va">TMB_per_Mb</span>,
                     corr.coeff <span class="op">=</span> <span class="st">"spearman"</span>,
                     title.plot<span class="op">=</span><span class="st">"Panel-WES correlation"</span><span class="op">)</span></code></pre></div>
<p><img src="Introduction_to_TMBleR_files/figure-html/unnamed-chunk-21-1.png" width="700"></p>
<p>In alternative, it is also possible to use the Pearson correlation coefficient by changing the <code>corr.coeff</code> argument in the function:</p>
<div class="sourceCode" id="cb23"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># Pearson correlation --------------------------------------------------------</span>
<span class="co"># Plot Pearson correlation</span>
<span class="co">#par(cex.axis=1.2, cex.lab=1.2)</span>
<span class="va">TMBs_SimulatedExamplePanel</span> <span class="op">%&gt;%</span>
  <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html">filter</a></span><span class="op">(</span><span class="va">Filter</span> <span class="op">==</span> <span class="st">"vaf=0NoCancerMuts"</span><span class="op">)</span> <span class="op">%&gt;%</span>
  <span class="va">.</span><span class="op">$</span><span class="va">TMB_per_Mb</span> <span class="op">%&gt;%</span>
<span class="fu"><a href="../reference/correlateTMBvalues.html">correlateTMBvalues</a></span><span class="op">(</span>panel.TMB <span class="op">=</span> ,
                   WES.TMB <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/numeric.html">as.numeric</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/vector.html">as.vector</a></span><span class="op">(</span><span class="va">TMBs_WES</span><span class="op">$</span><span class="va">TMB_per_Mb</span><span class="op">)</span><span class="op">)</span>,
                   corr.coeff <span class="op">=</span> <span class="st">"pearson"</span>,
                   title.plot<span class="op">=</span><span class="st">"Panel-WES correlation"</span><span class="op">)</span></code></pre></div>
<p><img src="Introduction_to_TMBleR_files/figure-html/unnamed-chunk-22-1.png" width="700"></p>
</div>
</div>
<div id="partC" class="section level1">
<h1 class="hasAnchor">
<a href="#partC" class="anchor"></a>Evaluate TMB performance to predict immunotherapy responders</h1>
<p>If both TMB and clinical data are available, the TMBleR package allows to build ROC curves and evaluate TMB performance in predicting immunotherapy responders. Here we use a dataset from Hellman et al., Cancer Cell 2018 (PMID: 29731394), in a study on immunotherapy response on small cell lung cancer patients. The original dataset provided as supplementary table contains tumor mutational burden, as total number of non-synonymous mutations identified by WES, and clinical response, which was transformed in a binary vector of values ‘responder’ and ‘nonresponder’ to be suitable to generateBoxplots and generateROC functions.</p>
<div id="dataset-preparation-2" class="section level3">
<h3 class="hasAnchor">
<a href="#dataset-preparation-2" class="anchor"></a>Dataset preparation</h3>
<p>There are 3 inputs required to perform the analysis:</p>
<ul>
<li>Genome assembly of reference: either “hg19” or “hg38”</li>
<li>TMB values</li>
<li>response to immunotherapy</li>
</ul>
<div id="genome-assembly-2" class="section level4">
<h4 class="hasAnchor">
<a href="#genome-assembly-2" class="anchor"></a>Genome assembly</h4>
<div class="sourceCode" id="cb24"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># 1. Specify human assembly of reference, either "hg19" or "hg38"</span>
<span class="co"># ==============================================================================</span>
<span class="va">genome_assembly</span> <span class="op">&lt;-</span> <span class="st">"hg19"</span></code></pre></div>
</div>
<div id="tmb-values" class="section level4">
<h4 class="hasAnchor">
<a href="#tmb-values" class="anchor"></a>TMB values</h4>
<p>Use in input TMB values generated above on simulated panel-based TMB already computed <code>TMB_res_df</code>.</p>
</div>
<div id="clinical-data" class="section level4">
<h4 class="hasAnchor">
<a href="#clinical-data" class="anchor"></a>Clinical data</h4>
<p>In this section we will import the clinical dataset published by <strong>VanAllen EM. et al.</strong> in Science 2015. This dataset contains lots of information but we only interest into knowning if the patient was considered as either a <strong>responder</strong> or <strong>nonresponder</strong>.</p>
<div class="sourceCode" id="cb25"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># 2. =========================================================================</span>
<span class="co"># Load dataset with clinical response to immunotherapy </span>
<span class="fu"><a href="https://rdrr.io/r/utils/data.html">data</a></span><span class="op">(</span><span class="st">"VanAllen_Clinical"</span><span class="op">)</span></code></pre></div>
<div id="annotate-tmb-values-with-clinical-data" class="section level5">
<h5 class="hasAnchor">
<a href="#annotate-tmb-values-with-clinical-data" class="anchor"></a>Annotate TMB values with clinical data</h5>
<div class="sourceCode" id="cb26"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">TMB_clinical</span> <span class="op">=</span> <span class="fu"><a href="../reference/annotateTMB.html">annotateTMB</a></span><span class="op">(</span>TMB_df <span class="op">=</span> <span class="va">TMB_res_df</span>, ClinicalData <span class="op">=</span> <span class="va">VanAllen_Clinical</span><span class="op">)</span></code></pre></div>
</div>
</div>
<div id="compare-tmb-distribution-between-responders-and-non-responders" class="section level4">
<h4 class="hasAnchor">
<a href="#compare-tmb-distribution-between-responders-and-non-responders" class="anchor"></a>Compare TMB distribution between responders and non-responders</h4>
<div class="sourceCode" id="cb27"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="../reference/generateBoxplot.html">generateBoxplot</a></span><span class="op">(</span><span class="va">TMB_clinical</span><span class="op">)</span></code></pre></div>
<p><img src="Introduction_to_TMBleR_files/figure-html/unnamed-chunk-26-1.png" width="700"></p>
</div>
<div id="perform-roc-analysis-to-evaluate-sensitivity-and-specificity-of-tmb-to-classify-responders-versus-non-responders-using-wes-based-tmb-values" class="section level4">
<h4 class="hasAnchor">
<a href="#perform-roc-analysis-to-evaluate-sensitivity-and-specificity-of-tmb-to-classify-responders-versus-non-responders-using-wes-based-tmb-values" class="anchor"></a>Perform ROC analysis to evaluate sensitivity and specificity of TMB to classify responders versus non-responders using WES-based TMB values</h4>
<p>The generateROC() function separates data in training (random 75% of input data) and test (remaining 25% of input data) using the caret package and provides metrics of classification performance using functions from the OptimalCutpoints package. It provides in output a list of objects containing all analyses results.</p>
<p>First, it identifies the best TMB cutoff, based on responders vs non-responders classification in the training set and provides in output: - best TMB cutoff, AUC, specificity, sensitivity, positive and negative predictive values, in the optimalCutpoints object “train_res” - the ROC curve, the predictive ROC curve and the variation of the optimal criterion value (i.e. Youden) for different cutoffs in the training set, in the optimalCutpoints object “train_res”</p>
<p>Then, it uses the identified TMB cutoff to perform classification of samples from the test set and provides in output: - the confusion matrix of test samples classification based on the best TMB cutoff value, in the confusionMatrix object “confMatrix_test” - the ROC curve for the test set, in the optimalCutpoints object “test_res”</p>
<div class="sourceCode" id="cb28"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># Set the seed to create reproducible train and test sets in generateROC</span>
<span class="fu"><a href="https://rdrr.io/r/base/Random.html">set.seed</a></span><span class="op">(</span><span class="fl">949</span><span class="op">)</span>


<span class="fu"><a href="https://rdrr.io/r/utils/data.html">data</a></span><span class="op">(</span><span class="va">Hellman_SimulatedFM1Panel_WES</span><span class="op">)</span>

<span class="co"># Fix dataset format</span>
<span class="va">Hellman_Df</span> <span class="op">&lt;-</span> <span class="va">Hellman_SimulatedFM1Panel_WES</span> <span class="op">%&gt;%</span>
  <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/rename.html">rename</a></span><span class="op">(</span><span class="st">"TMB_per_Mb"</span> <span class="op">=</span> <span class="st">"WES.NumMuts"</span><span class="op">)</span> <span class="op">%&gt;%</span>
  <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate.html">mutate</a></span><span class="op">(</span><span class="st">"ClinicalResponse"</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/factor.html">as.factor</a></span><span class="op">(</span><span class="va">.data</span><span class="op">$</span><span class="va">ClinicalResponse</span><span class="op">)</span><span class="op">)</span> <span class="op">%&gt;%</span>
  <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html">filter</a></span><span class="op">(</span><span class="op">!</span><span class="fu"><a href="https://rdrr.io/r/base/NA.html">is.na</a></span><span class="op">(</span><span class="va">.data</span><span class="op">$</span><span class="va">TMB_per_Mb</span><span class="op">)</span><span class="op">)</span> <span class="op">%&gt;%</span>
  <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html">filter</a></span><span class="op">(</span><span class="op">!</span><span class="fu"><a href="https://rdrr.io/r/base/NA.html">is.na</a></span><span class="op">(</span><span class="va">.data</span><span class="op">$</span><span class="va">ClinicalResponse</span><span class="op">)</span><span class="op">)</span>

    
<span class="co"># Analysis of  WES-based TMB performance in classification -------------------</span>
<span class="va">WES_ROCanalysis</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/generateROC.html">generateROC</a></span><span class="op">(</span>dataset <span class="op">=</span> <span class="va">Hellman_Df</span>,
                               method <span class="op">=</span> <span class="st">"Youden"</span><span class="op">)</span>

<span class="co"># Print AUC, sensitivity and specificity, calculated on training set</span>
<span class="fu"><a href="https://rdrr.io/r/base/print.html">print</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/summary.html">summary</a></span><span class="op">(</span><span class="va">WES_ROCanalysis</span><span class="op">$</span><span class="va">train_res</span><span class="op">)</span><span class="op">)</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; Call:</span>
<span class="co">#&gt; optimal.cutpoints.default(X = "TMB_per_Mb", status = "ClinicalResponse", </span>
<span class="co">#&gt;     tag.healthy = "nonresponder", methods = method, data = training, </span>
<span class="co">#&gt;     pop.prev = NULL, control = OptimalCutpoints::control.cutpoints(), </span>
<span class="co">#&gt;     ci.fit = TRUE, conf.level = 0.95, trace = FALSE)</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; Area under the ROC curve (AUC):  0.691 (0.525, 0.857) </span>
<span class="co">#&gt; </span>
<span class="co">#&gt; CRITERION: Youden</span>
<span class="co">#&gt; Number of optimal cutoffs: 1</span>
<span class="co">#&gt; </span>
<span class="co">#&gt;                      Estimate 95% CI lower limit 95% CI upper limit</span>
<span class="co">#&gt; cutoff            565.0000000                  -                  -</span>
<span class="co">#&gt; Se                  0.6666667          0.4099252          0.8665726</span>
<span class="co">#&gt; Sp                  0.7777778          0.6084824          0.8988495</span>
<span class="co">#&gt; PPV                 0.6000000          0.3997851          0.8296723</span>
<span class="co">#&gt; NPV                 0.8235294          0.6184616          0.9221689</span>
<span class="co">#&gt; DLR.Positive        3.0000000          1.5002929          5.9988285</span>
<span class="co">#&gt; DLR.Negative        0.4285714          0.2179369          0.8427828</span>
<span class="co">#&gt; FP                  8.0000000                  -                  -</span>
<span class="co">#&gt; FN                  6.0000000                  -                  -</span>
<span class="co">#&gt; Optimal criterion   0.4444444                  -                  -</span>

<span class="co"># Print confusion matrix describing the performance of classification on the</span>
<span class="co"># test set using the best TMB cutoff identified on the training set</span>
<span class="fu"><a href="https://rdrr.io/r/base/print.html">print</a></span><span class="op">(</span><span class="va">WES_ROCanalysis</span><span class="op">$</span><span class="va">confMatrix_test</span><span class="op">)</span>
<span class="co">#&gt; Confusion Matrix and Statistics</span>
<span class="co">#&gt; </span>
<span class="co">#&gt;               Reference</span>
<span class="co">#&gt; Prediction     responder nonresponder</span>
<span class="co">#&gt;   responder            3            3</span>
<span class="co">#&gt;   nonresponder         2            9</span>
<span class="co">#&gt;                                           </span>
<span class="co">#&gt;                Accuracy : 0.7059          </span>
<span class="co">#&gt;                  95% CI : (0.4404, 0.8969)</span>
<span class="co">#&gt;     No Information Rate : 0.7059          </span>
<span class="co">#&gt;     P-Value [Acc &gt; NIR] : 0.6177          </span>
<span class="co">#&gt;                                           </span>
<span class="co">#&gt;                   Kappa : 0.3307          </span>
<span class="co">#&gt;                                           </span>
<span class="co">#&gt;  Mcnemar's Test P-Value : 1.0000          </span>
<span class="co">#&gt;                                           </span>
<span class="co">#&gt;             Sensitivity : 0.6000          </span>
<span class="co">#&gt;             Specificity : 0.7500          </span>
<span class="co">#&gt;          Pos Pred Value : 0.5000          </span>
<span class="co">#&gt;          Neg Pred Value : 0.8182          </span>
<span class="co">#&gt;              Prevalence : 0.2941          </span>
<span class="co">#&gt;          Detection Rate : 0.1765          </span>
<span class="co">#&gt;    Detection Prevalence : 0.3529          </span>
<span class="co">#&gt;       Balanced Accuracy : 0.6750          </span>
<span class="co">#&gt;                                           </span>
<span class="co">#&gt;        'Positive' Class : responder       </span>
<span class="co">#&gt; </span>

<span class="co"># Plot ROC curve on test set</span>
<span class="fu"><a href="https://rdrr.io/r/graphics/par.html">par</a></span><span class="op">(</span>cex.axis<span class="op">=</span><span class="fl">1.5</span>, cex.lab<span class="op">=</span><span class="fl">1.5</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/graphics/plot.html">plot</a></span><span class="op">(</span><span class="va">WES_ROCanalysis</span><span class="op">$</span><span class="va">test_res</span>, which<span class="op">=</span><span class="fl">1</span><span class="op">)</span></code></pre></div>
<p><img src="Introduction_to_TMBleR_files/figure-html/unnamed-chunk-27-1.png" width="1344"></p>
<div class="sourceCode" id="cb29"><pre class="downlit sourceCode r">
<code class="sourceCode R">
<span class="co"># Analysis of  panel-based TMB performance in classification -------------------</span>
<span class="va">panel_ROCanalysis</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/generateROC.html">generateROC</a></span><span class="op">(</span>dataset <span class="op">=</span> <span class="va">Hellman_Df</span>,
                                 method <span class="op">=</span> <span class="st">"Youden"</span><span class="op">)</span>

<span class="co"># Print AUC, sensitivity and specificity, calculated on training set</span>
<span class="fu"><a href="https://rdrr.io/r/base/print.html">print</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/summary.html">summary</a></span><span class="op">(</span><span class="va">panel_ROCanalysis</span><span class="op">$</span><span class="va">train_res</span><span class="op">)</span><span class="op">)</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; Call:</span>
<span class="co">#&gt; optimal.cutpoints.default(X = "TMB_per_Mb", status = "ClinicalResponse", </span>
<span class="co">#&gt;     tag.healthy = "nonresponder", methods = method, data = training, </span>
<span class="co">#&gt;     pop.prev = NULL, control = OptimalCutpoints::control.cutpoints(), </span>
<span class="co">#&gt;     ci.fit = TRUE, conf.level = 0.95, trace = FALSE)</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; Area under the ROC curve (AUC):  0.663 (0.481, 0.845) </span>
<span class="co">#&gt; </span>
<span class="co">#&gt; CRITERION: Youden</span>
<span class="co">#&gt; Number of optimal cutoffs: 1</span>
<span class="co">#&gt; </span>
<span class="co">#&gt;                      Estimate 95% CI lower limit 95% CI upper limit</span>
<span class="co">#&gt; cutoff            565.0000000                  -                  -</span>
<span class="co">#&gt; Se                  0.6666667          0.4099252          0.8665726</span>
<span class="co">#&gt; Sp                  0.8055556          0.6397520          0.9180564</span>
<span class="co">#&gt; PPV                 0.6315789          0.4235782          0.8477211</span>
<span class="co">#&gt; NPV                 0.8285714          0.6267068          0.9289306</span>
<span class="co">#&gt; DLR.Positive        3.4285714          1.6345163          7.1917924</span>
<span class="co">#&gt; DLR.Negative        0.4137931          0.2111611          0.8108726</span>
<span class="co">#&gt; FP                  7.0000000                  -                  -</span>
<span class="co">#&gt; FN                  6.0000000                  -                  -</span>
<span class="co">#&gt; Optimal criterion   0.4722222                  -                  -</span>

<span class="co"># Plot ROC curve, PROC curve and plot of optimal criterion values depending on</span>
<span class="co"># TMB cutoff, calculated on the training set</span>
<span class="fu"><a href="https://rdrr.io/r/graphics/par.html">par</a></span><span class="op">(</span>mfrow<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">2</span>,<span class="fl">2</span><span class="op">)</span>, cex.axis<span class="op">=</span><span class="fl">1.5</span>, cex.lab<span class="op">=</span><span class="fl">1.5</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/graphics/plot.html">plot</a></span><span class="op">(</span><span class="va">panel_ROCanalysis</span><span class="op">$</span><span class="va">train_res</span>, which<span class="op">=</span><span class="fl">1</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/graphics/plot.html">plot</a></span><span class="op">(</span><span class="va">panel_ROCanalysis</span><span class="op">$</span><span class="va">train_res</span>, which<span class="op">=</span><span class="fl">2</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/graphics/plot.html">plot</a></span><span class="op">(</span><span class="va">panel_ROCanalysis</span><span class="op">$</span><span class="va">train_res</span>, which<span class="op">=</span><span class="fl">3</span>, ylim <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">0</span>, <span class="fl">1</span><span class="op">)</span><span class="op">)</span>

<span class="co"># Print confusion matrix describing the performance of classification on the</span>
<span class="co"># test set using the best TMB cutoff identified on the training set</span>
<span class="fu"><a href="https://rdrr.io/r/base/print.html">print</a></span><span class="op">(</span><span class="va">panel_ROCanalysis</span><span class="op">$</span><span class="va">confMatrix_test</span><span class="op">)</span>
<span class="co">#&gt; Confusion Matrix and Statistics</span>
<span class="co">#&gt; </span>
<span class="co">#&gt;               Reference</span>
<span class="co">#&gt; Prediction     responder nonresponder</span>
<span class="co">#&gt;   responder            3            4</span>
<span class="co">#&gt;   nonresponder         2            8</span>
<span class="co">#&gt;                                           </span>
<span class="co">#&gt;                Accuracy : 0.6471          </span>
<span class="co">#&gt;                  95% CI : (0.3833, 0.8579)</span>
<span class="co">#&gt;     No Information Rate : 0.7059          </span>
<span class="co">#&gt;     P-Value [Acc &gt; NIR] : 0.7914          </span>
<span class="co">#&gt;                                           </span>
<span class="co">#&gt;                   Kappa : 0.2388          </span>
<span class="co">#&gt;                                           </span>
<span class="co">#&gt;  Mcnemar's Test P-Value : 0.6831          </span>
<span class="co">#&gt;                                           </span>
<span class="co">#&gt;             Sensitivity : 0.6000          </span>
<span class="co">#&gt;             Specificity : 0.6667          </span>
<span class="co">#&gt;          Pos Pred Value : 0.4286          </span>
<span class="co">#&gt;          Neg Pred Value : 0.8000          </span>
<span class="co">#&gt;              Prevalence : 0.2941          </span>
<span class="co">#&gt;          Detection Rate : 0.1765          </span>
<span class="co">#&gt;    Detection Prevalence : 0.4118          </span>
<span class="co">#&gt;       Balanced Accuracy : 0.6333          </span>
<span class="co">#&gt;                                           </span>
<span class="co">#&gt;        'Positive' Class : responder       </span>
<span class="co">#&gt; </span>

<span class="co"># Plot ROC curve on test set</span>
<span class="fu"><a href="https://rdrr.io/r/graphics/par.html">par</a></span><span class="op">(</span>cex.axis<span class="op">=</span><span class="fl">1.5</span>, cex.lab<span class="op">=</span><span class="fl">1.5</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/graphics/plot.html">plot</a></span><span class="op">(</span><span class="va">panel_ROCanalysis</span><span class="op">$</span><span class="va">test_res</span>, which<span class="op">=</span><span class="fl">1</span><span class="op">)</span></code></pre></div>
<p><img src="Introduction_to_TMBleR_files/figure-html/unnamed-chunk-27-2.png" width="1344"></p>
</div>
</div>
<div id="compare-the-roc-curves-representing-the-performance-of-wes-based-and-panel-based-tmb-to-predict-clinical-response-" class="section level3">
<h3 class="hasAnchor">
<a href="#compare-the-roc-curves-representing-the-performance-of-wes-based-and-panel-based-tmb-to-predict-clinical-response-" class="anchor"></a>Compare the ROC curves representing the performance of WES-based and panel-based TMB to predict clinical response.</h3>
<div class="sourceCode" id="cb30"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/graphics/par.html">par</a></span><span class="op">(</span>cex.axis<span class="op">=</span><span class="fl">1.5</span>, cex.lab<span class="op">=</span><span class="fl">1.5</span><span class="op">)</span>
<span class="fu"><a href="../reference/compareROC.html">compareROC</a></span><span class="op">(</span>dataset <span class="op">=</span> <span class="va">Hellman_SimulatedFM1Panel_WES</span>,
           TMB1 <span class="op">=</span> <span class="st">"Panel.NumMuts"</span>,
           TMB2 <span class="op">=</span> <span class="st">"WES.NumMuts"</span>,
           method <span class="op">=</span> <span class="st">"d"</span>,
           paired <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span>
<span class="co">#&gt; Setting levels: control = nonresponder, case = responder</span>
<span class="co">#&gt; Setting direction: controls &lt; cases</span>
<span class="co">#&gt; Setting levels: control = nonresponder, case = responder</span>
<span class="co">#&gt; Setting direction: controls &lt; cases</span>
<span class="co">#&gt; </span>
<span class="co">#&gt;  DeLong's test for two correlated ROC curves</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; data:  TMB1 and TMB2 by as.factor(dataset$ClinicalResponse) (nonresponder, responder)</span>
<span class="co">#&gt; Z = 1.05, p-value = 0.2937</span>
<span class="co">#&gt; alternative hypothesis: true difference in AUC is not equal to 0</span>
<span class="co">#&gt; sample estimates:</span>
<span class="co">#&gt; AUC of roc1 AUC of roc2 </span>
<span class="co">#&gt;   0.7368659   0.6644022</span>
<span class="co">#&gt; Setting direction: controls &lt; cases</span>
<span class="co">#&gt; Setting direction: controls &lt; cases</span></code></pre></div>
<p><img src="Introduction_to_TMBleR_files/figure-html/unnamed-chunk-28-1.png" width="1152"></p>
<div class="sourceCode" id="cb31"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/utils/sessionInfo.html">sessionInfo</a></span><span class="op">(</span><span class="op">)</span>
<span class="co">#&gt; R version 3.6.3 (2020-02-29)</span>
<span class="co">#&gt; Platform: x86_64-apple-darwin15.6.0 (64-bit)</span>
<span class="co">#&gt; Running under: macOS  10.16</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; Matrix products: default</span>
<span class="co">#&gt; BLAS:   /Library/Frameworks/R.framework/Versions/3.6/Resources/lib/libRblas.0.dylib</span>
<span class="co">#&gt; LAPACK: /Library/Frameworks/R.framework/Versions/3.6/Resources/lib/libRlapack.dylib</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; locale:</span>
<span class="co">#&gt; [1] en_CA.UTF-8/en_CA.UTF-8/en_CA.UTF-8/C/en_CA.UTF-8/en_CA.UTF-8</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; attached base packages:</span>
<span class="co">#&gt; [1] stats     graphics  grDevices utils     datasets  methods   base     </span>
<span class="co">#&gt; </span>
<span class="co">#&gt; other attached packages:</span>
<span class="co">#&gt; [1] purrr_0.3.4      TMBleR_0.1.0     testthat_2.3.2   devtools_2.3.0  </span>
<span class="co">#&gt; [5] usethis_1.6.1    BiocStyle_2.12.0</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; loaded via a namespace (and not attached):</span>
<span class="co">#&gt;   [1] readxl_1.3.1                            </span>
<span class="co">#&gt;   [2] backports_1.1.7                         </span>
<span class="co">#&gt;   [3] BiocFileCache_1.10.2                    </span>
<span class="co">#&gt;   [4] systemfonts_1.0.1                       </span>
<span class="co">#&gt;   [5] plyr_1.8.6                              </span>
<span class="co">#&gt;   [6] splines_3.6.3                           </span>
<span class="co">#&gt;   [7] crosstalk_1.1.0.1                       </span>
<span class="co">#&gt;   [8] BiocParallel_1.20.1                     </span>
<span class="co">#&gt;   [9] GenomeInfoDb_1.22.1                     </span>
<span class="co">#&gt;  [10] ggplot2_3.3.1                           </span>
<span class="co">#&gt;  [11] digest_0.6.25                           </span>
<span class="co">#&gt;  [12] foreach_1.5.0                           </span>
<span class="co">#&gt;  [13] htmltools_0.5.0                         </span>
<span class="co">#&gt;  [14] magick_2.3                              </span>
<span class="co">#&gt;  [15] fansi_0.4.1                             </span>
<span class="co">#&gt;  [16] magrittr_1.5                            </span>
<span class="co">#&gt;  [17] memoise_1.1.0                           </span>
<span class="co">#&gt;  [18] BSgenome_1.54.0                         </span>
<span class="co">#&gt;  [19] openxlsx_4.1.5                          </span>
<span class="co">#&gt;  [20] remotes_2.1.1                           </span>
<span class="co">#&gt;  [21] recipes_0.1.12                          </span>
<span class="co">#&gt;  [22] Biostrings_2.54.0                       </span>
<span class="co">#&gt;  [23] gower_0.2.1                             </span>
<span class="co">#&gt;  [24] matrixStats_0.56.0                      </span>
<span class="co">#&gt;  [25] BSgenome.Hsapiens.UCSC.hg38_1.4.1       </span>
<span class="co">#&gt;  [26] askpass_1.1                             </span>
<span class="co">#&gt;  [27] pkgdown_1.6.1                           </span>
<span class="co">#&gt;  [28] prettyunits_1.1.1                       </span>
<span class="co">#&gt;  [29] colorspace_1.4-1                        </span>
<span class="co">#&gt;  [30] blob_1.2.1                              </span>
<span class="co">#&gt;  [31] rappdirs_0.3.1                          </span>
<span class="co">#&gt;  [32] textshaping_0.3.4                       </span>
<span class="co">#&gt;  [33] haven_2.3.1                             </span>
<span class="co">#&gt;  [34] xfun_0.18                               </span>
<span class="co">#&gt;  [35] dplyr_1.0.0                             </span>
<span class="co">#&gt;  [36] jsonlite_1.7.1                          </span>
<span class="co">#&gt;  [37] callr_3.4.3                             </span>
<span class="co">#&gt;  [38] crayon_1.3.4                            </span>
<span class="co">#&gt;  [39] RCurl_1.98-1.2                          </span>
<span class="co">#&gt;  [40] TxDb.Hsapiens.UCSC.hg19.knownGene_3.2.2 </span>
<span class="co">#&gt;  [41] survival_3.1-12                         </span>
<span class="co">#&gt;  [42] VariantAnnotation_1.30.1                </span>
<span class="co">#&gt;  [43] iterators_1.0.12                        </span>
<span class="co">#&gt;  [44] glue_1.4.2                              </span>
<span class="co">#&gt;  [45] gtable_0.3.0                            </span>
<span class="co">#&gt;  [46] ipred_0.9-9                             </span>
<span class="co">#&gt;  [47] zlibbioc_1.32.0                         </span>
<span class="co">#&gt;  [48] XVector_0.26.0                          </span>
<span class="co">#&gt;  [49] DelayedArray_0.12.3                     </span>
<span class="co">#&gt;  [50] TxDb.Hsapiens.UCSC.hg38.knownGene_3.10.0</span>
<span class="co">#&gt;  [51] car_3.0-8                               </span>
<span class="co">#&gt;  [52] pkgbuild_1.0.8                          </span>
<span class="co">#&gt;  [53] OptimalCutpoints_1.1-4                  </span>
<span class="co">#&gt;  [54] BiocGenerics_0.32.0                     </span>
<span class="co">#&gt;  [55] abind_1.4-5                             </span>
<span class="co">#&gt;  [56] scales_1.1.1                            </span>
<span class="co">#&gt;  [57] DBI_1.1.0                               </span>
<span class="co">#&gt;  [58] rstatix_0.5.0                           </span>
<span class="co">#&gt;  [59] Rcpp_1.0.5                              </span>
<span class="co">#&gt;  [60] progress_1.2.2                          </span>
<span class="co">#&gt;  [61] foreign_0.8-75                          </span>
<span class="co">#&gt;  [62] bit_1.1-15.2                            </span>
<span class="co">#&gt;  [63] lava_1.6.7                              </span>
<span class="co">#&gt;  [64] prodlim_2019.11.13                      </span>
<span class="co">#&gt;  [65] stats4_3.6.3                            </span>
<span class="co">#&gt;  [66] DT_0.13                                 </span>
<span class="co">#&gt;  [67] htmlwidgets_1.5.3                       </span>
<span class="co">#&gt;  [68] httr_1.4.2                              </span>
<span class="co">#&gt;  [69] RColorBrewer_1.1-2                      </span>
<span class="co">#&gt;  [70] ellipsis_0.3.1                          </span>
<span class="co">#&gt;  [71] farver_2.0.3                            </span>
<span class="co">#&gt;  [72] pkgconfig_2.0.3                         </span>
<span class="co">#&gt;  [73] XML_3.99-0.3                            </span>
<span class="co">#&gt;  [74] nnet_7.3-14                             </span>
<span class="co">#&gt;  [75] dbplyr_1.4.4                            </span>
<span class="co">#&gt;  [76] caret_6.0-86                            </span>
<span class="co">#&gt;  [77] labeling_0.3                            </span>
<span class="co">#&gt;  [78] reshape2_1.4.4                          </span>
<span class="co">#&gt;  [79] tidyselect_1.1.0                        </span>
<span class="co">#&gt;  [80] rlang_0.4.7                             </span>
<span class="co">#&gt;  [81] AnnotationDbi_1.48.0                    </span>
<span class="co">#&gt;  [82] munsell_0.5.0                           </span>
<span class="co">#&gt;  [83] cellranger_1.1.0                        </span>
<span class="co">#&gt;  [84] tools_3.6.3                             </span>
<span class="co">#&gt;  [85] cli_2.0.2                               </span>
<span class="co">#&gt;  [86] generics_0.0.2                          </span>
<span class="co">#&gt;  [87] RSQLite_2.2.0                           </span>
<span class="co">#&gt;  [88] broom_0.7.2                             </span>
<span class="co">#&gt;  [89] evaluate_0.14                           </span>
<span class="co">#&gt;  [90] stringr_1.4.0                           </span>
<span class="co">#&gt;  [91] yaml_2.2.1                              </span>
<span class="co">#&gt;  [92] ragg_1.1.2                              </span>
<span class="co">#&gt;  [93] ModelMetrics_1.2.2.2                    </span>
<span class="co">#&gt;  [94] processx_3.4.2                          </span>
<span class="co">#&gt;  [95] knitr_1.30                              </span>
<span class="co">#&gt;  [96] bit64_0.9-7                             </span>
<span class="co">#&gt;  [97] fs_1.4.1                                </span>
<span class="co">#&gt;  [98] zip_2.0.4                               </span>
<span class="co">#&gt;  [99] nlme_3.1-148                            </span>
<span class="co">#&gt; [100] biomaRt_2.42.1                          </span>
<span class="co">#&gt; [101] compiler_3.6.3                          </span>
<span class="co">#&gt; [102] rstudioapi_0.11                         </span>
<span class="co">#&gt; [103] curl_4.3                                </span>
<span class="co">#&gt; [104] e1071_1.7-3                             </span>
<span class="co">#&gt; [105] ggsignif_0.6.0                          </span>
<span class="co">#&gt; [106] tibble_3.0.1                            </span>
<span class="co">#&gt; [107] stringi_1.4.6                           </span>
<span class="co">#&gt; [108] ps_1.3.3                                </span>
<span class="co">#&gt; [109] GenomicFeatures_1.38.2                  </span>
<span class="co">#&gt; [110] desc_1.2.0                              </span>
<span class="co">#&gt; [111] forcats_0.5.0                           </span>
<span class="co">#&gt; [112] lattice_0.20-41                         </span>
<span class="co">#&gt; [113] Matrix_1.2-18                           </span>
<span class="co">#&gt; [114] vctrs_0.3.4                             </span>
<span class="co">#&gt; [115] pillar_1.4.4                            </span>
<span class="co">#&gt; [116] lifecycle_0.2.0                         </span>
<span class="co">#&gt; [117] BiocManager_1.30.10                     </span>
<span class="co">#&gt; [118] data.table_1.12.8                       </span>
<span class="co">#&gt; [119] bitops_1.0-6                            </span>
<span class="co">#&gt; [120] rtracklayer_1.46.0                      </span>
<span class="co">#&gt; [121] GenomicRanges_1.38.0                    </span>
<span class="co">#&gt; [122] R6_2.4.1                                </span>
<span class="co">#&gt; [123] rio_0.5.16                              </span>
<span class="co">#&gt; [124] IRanges_2.20.2                          </span>
<span class="co">#&gt; [125] sessioninfo_1.1.1                       </span>
<span class="co">#&gt; [126] codetools_0.2-16                        </span>
<span class="co">#&gt; [127] MASS_7.3-51.6                           </span>
<span class="co">#&gt; [128] assertthat_0.2.1                        </span>
<span class="co">#&gt; [129] pkgload_1.1.0                           </span>
<span class="co">#&gt; [130] cutpointr_1.0.2                         </span>
<span class="co">#&gt; [131] SummarizedExperiment_1.16.1             </span>
<span class="co">#&gt; [132] openssl_1.4.1                           </span>
<span class="co">#&gt; [133] rprojroot_1.3-2                         </span>
<span class="co">#&gt; [134] withr_2.3.0                             </span>
<span class="co">#&gt; [135] GenomicAlignments_1.22.1                </span>
<span class="co">#&gt; [136] Rsamtools_2.2.3                         </span>
<span class="co">#&gt; [137] S4Vectors_0.24.4                        </span>
<span class="co">#&gt; [138] GenomeInfoDbData_1.2.2                  </span>
<span class="co">#&gt; [139] parallel_3.6.3                          </span>
<span class="co">#&gt; [140] hms_0.5.3                               </span>
<span class="co">#&gt; [141] rpart_4.1-15                            </span>
<span class="co">#&gt; [142] grid_3.6.3                              </span>
<span class="co">#&gt; [143] timeDate_3043.102                       </span>
<span class="co">#&gt; [144] tidyr_1.1.0                             </span>
<span class="co">#&gt; [145] class_7.3-17                            </span>
<span class="co">#&gt; [146] rmarkdown_2.6                           </span>
<span class="co">#&gt; [147] BSgenome.Hsapiens.UCSC.hg19_1.4.0       </span>
<span class="co">#&gt; [148] carData_3.0-4                           </span>
<span class="co">#&gt; [149] ggpubr_0.3.0                            </span>
<span class="co">#&gt; [150] pROC_1.16.2                             </span>
<span class="co">#&gt; [151] lubridate_1.7.9                         </span>
<span class="co">#&gt; [152] Biobase_2.46.0</span></code></pre></div>
</div>
</div>
<div class="footnotes">
<hr>
<ol>
<li id="fn1"><p>In the example showcased we are using targeted sequencing (specific group of genes), but the application of this package can be extended also to whole exome sequencing (WES).<a href="#fnref1" class="footnote-back">↩︎</a></p></li>
<li id="fn2"><p>(Horizon, provides cell lines with a known mutational provide as well as the variant allele frequency (VAF). The samples in this example, are both characterized by mutations at 5% VAF; one of them was exposed to mild Formalin-fixed and paraffin-embedded (FFPE) treatment to reproduce the effects of DNA degradation typical of FFPE).<a href="#fnref2" class="footnote-back">↩︎</a></p></li>
</ol>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

        <nav id="toc" data-toggle="toc"><h2 data-toc-skip>Contents</h2>
    </nav>
</div>

</div>



      <footer><div class="copyright">
  <p>Developed by Alessandro Guida, Laura Fancello, Luca Mazzarella.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.6.1.</p>
</div>

      </footer>
</div>

  


  </body>
</html>
